{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyP+N089BZGODxk1ALHXpAWT"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","metadata":{"id":"kqdNELCCOMPy"},"source":["**Script to obtain the data files that are used to initialize the simulations of transcriptional and translational divergence of yeast duplicated genes**"]},{"cell_type":"code","metadata":{"id":"nlxHPg2AO2tt","executionInfo":{"status":"ok","timestamp":1671477873311,"user_tz":300,"elapsed":136,"user":{"displayName":"Simon Aubé","userId":"10741788869764524869"}}},"source":["import numpy as np\n","import pandas as pd\n","from google.colab import files\n","import io\n","import scipy.stats as stats\n","import math"],"execution_count":31,"outputs":[]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"XtWuZ-afO_GO","executionInfo":{"status":"ok","timestamp":1671477874274,"user_tz":300,"elapsed":800,"user":{"displayName":"Simon Aubé","userId":"10741788869764524869"}},"outputId":"aeeb6ef1-659e-4843-a17d-9bb7c5b0b01c"},"source":["# Mounting my google drive\n","\n","from google.colab import drive\n","drive.mount('/content/drive')"],"execution_count":32,"outputs":[{"output_type":"stream","name":"stdout","text":["Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount(\"/content/drive\", force_remount=True).\n"]}]},{"cell_type":"markdown","metadata":{"id":"PJ72tXC9OmxI"},"source":["Transcription and translation rates from Hausser et al. (2019) are first imported."]},{"cell_type":"code","metadata":{"id":"5g2pbW6jOCyo","executionInfo":{"status":"ok","timestamp":1671477874274,"user_tz":300,"elapsed":4,"user":{"displayName":"Simon Aubé","userId":"10741788869764524869"}}},"source":["rates_yeast = pd.read_csv('/content/drive/My Drive/Redaction_SA/Figures_2021_06/Data_init/yeastRates.csv', sep=';', decimal=',')\n","rates_yeast = rates_yeast.rename(columns={'Unnamed: 0': 'Gene'})"],"execution_count":33,"outputs":[]},{"cell_type":"code","source":["rates_yeast"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":424},"id":"fAh7G8TaBmJT","executionInfo":{"status":"ok","timestamp":1671477874274,"user_tz":300,"elapsed":4,"user":{"displayName":"Simon Aubé","userId":"10741788869764524869"}},"outputId":"1ca67cbc-bfb0-4aa5-cc05-dc72e4e221e7"},"execution_count":34,"outputs":[{"output_type":"execute_result","data":{"text/plain":["       Gene        lm        lp      wRPF     wmRNA    bmEser    amEser  \\\n","0      TFC3  3.553033  3.064832  1.258158  1.577492  1.125383  0.704432   \n","1      VPS8  3.582631  3.105510  0.758155  1.345766  1.146577  0.868402   \n","2      EFB1  2.994317  2.315970  3.849116  3.664817  2.307282  0.058252   \n","3      ERP2  2.811575  2.334454  2.321122  2.327195  1.701383  0.057926   \n","4     FUN14  2.775974  2.298853  1.980231  1.698622  0.790021  0.382707   \n","...     ...       ...       ...       ...       ...       ...       ...   \n","4435   OPT2  3.420616  2.943495 -0.236572  0.842609 -0.399593  0.274546   \n","4436   SGE1  3.212720  2.735599  1.472025  1.521922  1.158760  0.532465   \n","4437   ARR1  2.946943  2.469822  1.138934  1.592954  1.449894  0.730236   \n","4438   ARR2  2.594393  2.117271  0.387390  1.169086  0.400387  0.863932   \n","4439   ARR3  3.084576  2.607455  0.305351  0.501059 -0.077695  0.703301   \n","\n","            bm         m        bp        cv     apExp        am  isEssential  \\\n","0     1.127728  0.423025  2.201503       NaN -1.057717  0.704302         True   \n","1     0.896002  0.191299  1.892547       NaN       NaN  0.704302        False   \n","2     3.215053  2.510350  3.453998       NaN       NaN  0.704302         True   \n","3     1.877432  1.172729  3.245141 -0.744004 -0.071219  0.704302        False   \n","4     1.248859  0.544156  3.568424       NaN  0.204003  0.704302        False   \n","...        ...       ...       ...       ...       ...       ...          ...   \n","4435  0.392846 -0.311857  1.562993       NaN       NaN  0.704302        False   \n","4436  1.072159  0.367456  2.800172       NaN       NaN  0.704302        False   \n","4437  1.143190  0.438487  2.661827       NaN       NaN  0.704302        False   \n","4438  0.719323  0.014620  2.686701       NaN       NaN  0.704302        False   \n","4439  0.051296 -0.653407  2.782506       NaN       NaN  0.704302        False   \n","\n","     hasTATA   YEPDFit      pEst  \n","0       True  0.040325  2.496036  \n","1      False  0.020725  1.955355  \n","2      False  0.420089  5.835856  \n","3      False -0.026295  4.289378  \n","4      False -0.018375  3.984088  \n","...      ...       ...       ...  \n","4435    True -0.009625  1.122644  \n","4436   False -0.009200  3.039136  \n","4437   False -0.016960  2.971822  \n","4438    True -0.010200  2.572829  \n","4439    True -0.008790  2.000607  \n","\n","[4440 rows x 17 columns]"],"text/html":["\n","  <div id=\"df-9cfcc3b1-fde5-46b7-af01-2f79b15d8b2c\">\n","    <div class=\"colab-df-container\">\n","      <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Gene</th>\n","      <th>lm</th>\n","      <th>lp</th>\n","      <th>wRPF</th>\n","      <th>wmRNA</th>\n","      <th>bmEser</th>\n","      <th>amEser</th>\n","      <th>bm</th>\n","      <th>m</th>\n","      <th>bp</th>\n","      <th>cv</th>\n","      <th>apExp</th>\n","      <th>am</th>\n","      <th>isEssential</th>\n","      <th>hasTATA</th>\n","      <th>YEPDFit</th>\n","      <th>pEst</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>TFC3</td>\n","      <td>3.553033</td>\n","      <td>3.064832</td>\n","      <td>1.258158</td>\n","      <td>1.577492</td>\n","      <td>1.125383</td>\n","      <td>0.704432</td>\n","      <td>1.127728</td>\n","      <td>0.423025</td>\n","      <td>2.201503</td>\n","      <td>NaN</td>\n","      <td>-1.057717</td>\n","      <td>0.704302</td>\n","      <td>True</td>\n","      <td>True</td>\n","      <td>0.040325</td>\n","      <td>2.496036</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>VPS8</td>\n","      <td>3.582631</td>\n","      <td>3.105510</td>\n","      <td>0.758155</td>\n","      <td>1.345766</td>\n","      <td>1.146577</td>\n","      <td>0.868402</td>\n","      <td>0.896002</td>\n","      <td>0.191299</td>\n","      <td>1.892547</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>0.704302</td>\n","      <td>False</td>\n","      <td>False</td>\n","      <td>0.020725</td>\n","      <td>1.955355</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>EFB1</td>\n","      <td>2.994317</td>\n","      <td>2.315970</td>\n","      <td>3.849116</td>\n","      <td>3.664817</td>\n","      <td>2.307282</td>\n","      <td>0.058252</td>\n","      <td>3.215053</td>\n","      <td>2.510350</td>\n","      <td>3.453998</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>0.704302</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>0.420089</td>\n","      <td>5.835856</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>ERP2</td>\n","      <td>2.811575</td>\n","      <td>2.334454</td>\n","      <td>2.321122</td>\n","      <td>2.327195</td>\n","      <td>1.701383</td>\n","      <td>0.057926</td>\n","      <td>1.877432</td>\n","      <td>1.172729</td>\n","      <td>3.245141</td>\n","      <td>-0.744004</td>\n","      <td>-0.071219</td>\n","      <td>0.704302</td>\n","      <td>False</td>\n","      <td>False</td>\n","      <td>-0.026295</td>\n","      <td>4.289378</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>FUN14</td>\n","      <td>2.775974</td>\n","      <td>2.298853</td>\n","      <td>1.980231</td>\n","      <td>1.698622</td>\n","      <td>0.790021</td>\n","      <td>0.382707</td>\n","      <td>1.248859</td>\n","      <td>0.544156</td>\n","      <td>3.568424</td>\n","      <td>NaN</td>\n","      <td>0.204003</td>\n","      <td>0.704302</td>\n","      <td>False</td>\n","      <td>False</td>\n","      <td>-0.018375</td>\n","      <td>3.984088</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>4435</th>\n","      <td>OPT2</td>\n","      <td>3.420616</td>\n","      <td>2.943495</td>\n","      <td>-0.236572</td>\n","      <td>0.842609</td>\n","      <td>-0.399593</td>\n","      <td>0.274546</td>\n","      <td>0.392846</td>\n","      <td>-0.311857</td>\n","      <td>1.562993</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>0.704302</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>-0.009625</td>\n","      <td>1.122644</td>\n","    </tr>\n","    <tr>\n","      <th>4436</th>\n","      <td>SGE1</td>\n","      <td>3.212720</td>\n","      <td>2.735599</td>\n","      <td>1.472025</td>\n","      <td>1.521922</td>\n","      <td>1.158760</td>\n","      <td>0.532465</td>\n","      <td>1.072159</td>\n","      <td>0.367456</td>\n","      <td>2.800172</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>0.704302</td>\n","      <td>False</td>\n","      <td>False</td>\n","      <td>-0.009200</td>\n","      <td>3.039136</td>\n","    </tr>\n","    <tr>\n","      <th>4437</th>\n","      <td>ARR1</td>\n","      <td>2.946943</td>\n","      <td>2.469822</td>\n","      <td>1.138934</td>\n","      <td>1.592954</td>\n","      <td>1.449894</td>\n","      <td>0.730236</td>\n","      <td>1.143190</td>\n","      <td>0.438487</td>\n","      <td>2.661827</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>0.704302</td>\n","      <td>False</td>\n","      <td>False</td>\n","      <td>-0.016960</td>\n","      <td>2.971822</td>\n","    </tr>\n","    <tr>\n","      <th>4438</th>\n","      <td>ARR2</td>\n","      <td>2.594393</td>\n","      <td>2.117271</td>\n","      <td>0.387390</td>\n","      <td>1.169086</td>\n","      <td>0.400387</td>\n","      <td>0.863932</td>\n","      <td>0.719323</td>\n","      <td>0.014620</td>\n","      <td>2.686701</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>0.704302</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>-0.010200</td>\n","      <td>2.572829</td>\n","    </tr>\n","    <tr>\n","      <th>4439</th>\n","      <td>ARR3</td>\n","      <td>3.084576</td>\n","      <td>2.607455</td>\n","      <td>0.305351</td>\n","      <td>0.501059</td>\n","      <td>-0.077695</td>\n","      <td>0.703301</td>\n","      <td>0.051296</td>\n","      <td>-0.653407</td>\n","      <td>2.782506</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>0.704302</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>-0.008790</td>\n","      <td>2.000607</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>4440 rows × 17 columns</p>\n","</div>\n","      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-9cfcc3b1-fde5-46b7-af01-2f79b15d8b2c')\"\n","              title=\"Convert this dataframe to an interactive table.\"\n","              style=\"display:none;\">\n","        \n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","       width=\"24px\">\n","    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n","    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n","  </svg>\n","      </button>\n","      \n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      flex-wrap:wrap;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","      <script>\n","        const buttonEl =\n","          document.querySelector('#df-9cfcc3b1-fde5-46b7-af01-2f79b15d8b2c button.colab-df-convert');\n","        buttonEl.style.display =\n","          google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","        async function convertToInteractive(key) {\n","          const element = document.querySelector('#df-9cfcc3b1-fde5-46b7-af01-2f79b15d8b2c');\n","          const dataTable =\n","            await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                     [key], {});\n","          if (!dataTable) return;\n","\n","          const docLinkHtml = 'Like what you see? Visit the ' +\n","            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","            + ' to learn more about interactive tables.';\n","          element.innerHTML = '';\n","          dataTable['output_type'] = 'display_data';\n","          await google.colab.output.renderOutput(dataTable, element);\n","          const docLink = document.createElement('div');\n","          docLink.innerHTML = docLinkHtml;\n","          element.appendChild(docLink);\n","        }\n","      </script>\n","    </div>\n","  </div>\n","  "]},"metadata":{},"execution_count":34}]},{"cell_type":"markdown","metadata":{"id":"ayF5mcI0WKrN"},"source":["## 1) Addition of systematic names"]},{"cell_type":"markdown","metadata":{"id":"-5cPWGM1bT69"},"source":["The file saccharomyces_cerevisiae.20200611.gff has been downloaded from Saccharomyces Genome Database (SGD). It contains all genomic features in the S288C reference genome, as of 2020-06-11, as well as the complete genome sequence. It is used to obtain all the systematic - standard name associations."]},{"cell_type":"code","metadata":{"id":"0md2sFErf4eq","executionInfo":{"status":"ok","timestamp":1671477874445,"user_tz":300,"elapsed":174,"user":{"displayName":"Simon Aubé","userId":"10741788869764524869"}}},"source":["features_SGD = pd.read_table('/content/drive/My Drive/Redaction_SA/Figures_2021_06/Data_init/saccharomyces_cerevisiae.20200611.gff', \n","                             skiprows=20, header=None, index_col=False, \n","                             names=['seqname', 'source', 'feature', 'start', \n","                                    'end', 'score', 'strand', 'frame', \n","                                    'attribute'], nrows=23067)\n","# The genome sequence is not read\n","\n","features_genes = features_SGD[features_SGD['feature'] == 'gene']\n","features_genes = features_genes.reset_index(drop=True)"],"execution_count":35,"outputs":[]},{"cell_type":"code","metadata":{"id":"jLGyWVkJhW3o","executionInfo":{"status":"ok","timestamp":1671477874633,"user_tz":300,"elapsed":189,"user":{"displayName":"Simon Aubé","userId":"10741788869764524869"}}},"source":["# The features_genes file is used to build a dictionary of standard name -\n","# systematic name associations\n","\n","genes_orf = {}\n","\n","for row in range(features_genes.shape[0]):\n","\n","  features = features_genes.at[row, 'attribute'].split(';')\n","  feat_dict = {}\n","\n","  for feat in features:\n","    feat_split = feat.split('=')\n","\n","    feat_dict[feat_split[0]] = feat_split[1]\n","\n","  if 'gene' in feat_dict.keys(): # For cases when the two IDs are the same \n","                                 # (ex: mitochondrial genes)\n","\n","    genes_orf[feat_dict['gene']] = feat_dict['ID']\n","\n","  else:\n","\n","    genes_orf[feat_dict['ID']] = feat_dict['ID']"],"execution_count":36,"outputs":[]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"iCmJM_y4iJEY","executionInfo":{"status":"ok","timestamp":1671477874633,"user_tz":300,"elapsed":3,"user":{"displayName":"Simon Aubé","userId":"10741788869764524869"}},"outputId":"2e95f759-d6ab-4aff-d32d-99182f1cc084"},"source":["# This dictionary is used to add the systematic ID of genes to the rates df\n","rates_yeast.insert(0, 'ORF', 'problem')\n","missing = []\n","\n","for row in range(rates_yeast.shape[0]):\n","  gene = rates_yeast.at[row, 'Gene']\n","\n","  if gene in genes_orf.keys():\n","    rates_yeast.at[row, 'ORF'] = genes_orf[gene]\n","  \n","  else:\n","\n","    missing.append(gene)\n","\n","print(missing)"],"execution_count":37,"outputs":[{"output_type":"stream","name":"stdout","text":["['AIM1', 'PHO88', 'DUR1,2', 'ARG5,6', 'FRA2', 'ADE5,7', 'AML1', 'YSP1', 'PET10', 'ENV10', 'MDN1', 'FLD1', 'OSW5', 'ICY2']\n"]}]},{"cell_type":"code","metadata":{"id":"0B5K0EYoiTqU","executionInfo":{"status":"ok","timestamp":1671477874634,"user_tz":300,"elapsed":3,"user":{"displayName":"Simon Aubé","userId":"10741788869764524869"}}},"source":["# For the missing genes, corresponding to alternative aliases, the systematic \n","# IDs are added manually\n","\n","missing_ID = {'AIM1': 'YAL046C', 'PHO88': 'YBR106W', 'DUR1,2': 'YBR208C', \n","              'ARG5,6': 'YER069W', 'FRA2': 'YGL220W', 'ADE5,7': 'YGL234W',\n","              'AML1': 'YGR001C', 'YSP1': 'YHR155W', 'PET10': 'YKR046C',\n","              'ENV10': 'YLR065C', 'MDN1': 'YLR106C', 'FLD1': 'YLR404W',\n","              'OSW5': 'YMR148W', 'ICY2': 'YPL250C'}\n","\n","for gene in missing:\n","\n","  row = rates_yeast[rates_yeast['Gene'] == gene].index\n","\n","  rates_yeast.at[row, 'ORF'] = missing_ID[gene]"],"execution_count":38,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"HL7zJb4BiwF7"},"source":["## 2) Addition of duplication-type annotations"]},{"cell_type":"markdown","metadata":{"id":"n4Hz9D7ki3eR"},"source":["Table S1 from Supplementary file 2 of Marchant et al. (2019) is imported. In this file, each yeast ORF (systematic ID) is associated to a duplication type annotation (S: singleton, SSD: small scale duplication,  SSD-successive: successive small-scale duplication, WGD: whole-genome duplication or SSD_WGD: paralogs duplicated by both mechanism of duplication SSD and WGD)."]},{"cell_type":"code","metadata":{"id":"LqwN8G2VjVeN","executionInfo":{"status":"ok","timestamp":1671477874762,"user_tz":300,"elapsed":131,"user":{"displayName":"Simon Aubé","userId":"10741788869764524869"}}},"source":["dupli_type = pd.read_csv('/content/drive/My Drive/Redaction_SA/Figures_2021_06/Data_init/TableS1_Marchant_2019.csv', sep=';')"],"execution_count":39,"outputs":[]},{"cell_type":"code","metadata":{"id":"f09jlRF3lfso","executionInfo":{"status":"ok","timestamp":1671477881787,"user_tz":300,"elapsed":7026,"user":{"displayName":"Simon Aubé","userId":"10741788869764524869"}}},"source":["rates_yeast['Duplication'] = 'Unspecified' # For cases that are not in\n","                                           # the dupli_type dataframe\n","\n","for row in range(rates_yeast.shape[0]):\n","  orf = rates_yeast.at[row, 'ORF']\n","\n","  if orf in list(dupli_type['ORF']):\n","    dupli = dupli_type[dupli_type['ORF'] == orf]['Duplication'].values[0]\n","    rates_yeast.at[row, 'Duplication'] = dupli"],"execution_count":40,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"g-ep7idVmbGC"},"source":["## 3) Recalculation of transcription and translation rates and rates_Hausser.csv export"]},{"cell_type":"markdown","source":["We recalculated all expression rates reported by Hausser et al. (2019) from the raw data deposited by Weinberg et al. This was done because we mistakenly thought that the calculation of translation rates by Hausser et al. was erroneous. "],"metadata":{"id":"Qtc3zEvzB6VE"}},{"cell_type":"markdown","source":["The recalculated translation rates are not used in our work, but some of the preliminary simulations were performed using both the original and recalculated (\"corrected\") sets of translations rates. In these, the choice of the dataset had no impact on the results."],"metadata":{"id":"I8rjttBvJgdz"}},{"cell_type":"markdown","source":["**Note:** The recalculated transcription rates have been used in the simulations, simply because the script was at first defined that way. They are equivalent to the rates originally reported by Hausser et al. (2019)."],"metadata":{"id":"na7jM0pny6_i"}},{"cell_type":"markdown","metadata":{"id":"kkAvs6l0mp-K"},"source":["Transcription and translation rates, as well as protein abundance, are recalculated from the RPKMs reported by Weinberg et al. (2016), following the Methods of Hausser et al. (2019)."]},{"cell_type":"markdown","metadata":{"id":"gIz9PQpcrDNr"},"source":["Raw data are downloaded from GSE75897 (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE75897). The RNA-seq RPKM from the RiboZero experiment (GSE75897_RiboZero_RPKMs.txt.gz) and the ribosome profiling RPKMs (GSE75897_RPF_RPKMs.txt.gz) are obtained and imported:"]},{"cell_type":"code","metadata":{"id":"qIImz056nEYW","executionInfo":{"status":"ok","timestamp":1671477881788,"user_tz":300,"elapsed":11,"user":{"displayName":"Simon Aubé","userId":"10741788869764524869"}}},"source":["mRNA_RPKM = pd.read_csv('/content/drive/My Drive/Redaction_SA/Figures_2021_06/Data_init/GSE75897_RiboZero_RPKMs.txt', sep='\\t', names=['ORF', 'mRNA RPKM'])\n","RP_RPKM = pd.read_csv('/content/drive/My Drive/Redaction_SA/Figures_2021_06/Data_init/GSE75897_RPF_RPKMs.txt', sep='\\t', names=['ORF', 'RP RPKM'])"],"execution_count":41,"outputs":[]},{"cell_type":"code","metadata":{"id":"xvyCQGOHsFHx","executionInfo":{"status":"ok","timestamp":1671477881788,"user_tz":300,"elapsed":10,"user":{"displayName":"Simon Aubé","userId":"10741788869764524869"}}},"source":["# To obtain a file with only genes for which RPKMs have been reported in\n","# both experiments\n","\n","RPKM_all = pd.merge(mRNA_RPKM, RP_RPKM, how='outer')\n","\n","RPKM_both = RPKM_all.dropna().reset_index(drop=True)\n","\n","# RPKM of 0.0 are also dropped\n","RPKM_both = RPKM_both[(RPKM_both['mRNA RPKM'] != 0) & (RPKM_both['RP RPKM'] != 0)].reset_index(drop=True)"],"execution_count":42,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"IwI27q9Ksc1B"},"source":["There are 5068 genes for which both types of RPKMs are available, compared with 4400 in the dataset from Hausser et al. (2019). This seems consistent with the filtering step described in the Methods of this paper."]},{"cell_type":"code","metadata":{"id":"OUShMhMRib9i","executionInfo":{"status":"ok","timestamp":1671477881788,"user_tz":300,"elapsed":10,"user":{"displayName":"Simon Aubé","userId":"10741788869764524869"}}},"source":["# Constants from Hausser et al. (2019)\n","alpha_m = 5.1\n","alpha_p = 1.34\n","n_mRNA = 60000\n","n_prot = 1.1e8\n","\n","# Transcription and translation rates are calculated (Hausser et al., 2019)\n","RPKM_both['m_est'] = np.log10(n_mRNA*(RPKM_both['mRNA RPKM']/np.sum(RPKM_both['mRNA RPKM'])))\n","RPKM_both['bm_calc'] = np.log10((10**RPKM_both['m_est']) * alpha_m)\n","\n","RPKM_both['bp_calc'] = np.log10(((n_prot * alpha_p)/(10**RPKM_both['m_est']))*((RPKM_both['RP RPKM'])/np.sum(RP_RPKM['RP RPKM'])))\n","\n","# As well as a new estimated protein abundance\n","RPKM_both['pEst_calc'] = np.log10(((10**RPKM_both['bm_calc']) * (10**RPKM_both['bp_calc'])) / (alpha_m * alpha_p))"],"execution_count":43,"outputs":[]},{"cell_type":"code","metadata":{"id":"tR9yJ8GsivHW","executionInfo":{"status":"ok","timestamp":1671477881789,"user_tz":300,"elapsed":10,"user":{"displayName":"Simon Aubé","userId":"10741788869764524869"}}},"source":["# Merging with the yeastRates dataframe \n","rates_corrected = pd.merge(rates_yeast, RPKM_both[['ORF', 'm_est', 'bm_calc', 'bp_calc', 'pEst_calc']], how='left', on=['ORF'])"],"execution_count":44,"outputs":[]},{"cell_type":"markdown","source":["**Most important:** Recalculation of the transcription rates from Hausser et al. to account for variation in mRNA decay rates, as measured by Eser et al. (2016):"],"metadata":{"id":"tXvAkezR5gxD"}},{"cell_type":"code","source":["# The decay rates from Eser et al. first need to be corrected for dilution due\n","# to cell division\n","rates_corrected['amEser'] = np.log10(10**rates_corrected['amEser'] + (math.log(2)/(99/60)))"],"metadata":{"id":"etaZRPqP1LCS","executionInfo":{"status":"ok","timestamp":1671477882017,"user_tz":300,"elapsed":238,"user":{"displayName":"Simon Aubé","userId":"10741788869764524869"}}},"execution_count":45,"outputs":[]},{"cell_type":"code","source":["rates_corrected['bm_Eser'] = np.log10(10**rates_corrected['m'] * 10**rates_corrected['amEser'])"],"metadata":{"id":"sDk7gs_z5093","executionInfo":{"status":"ok","timestamp":1671477882017,"user_tz":300,"elapsed":4,"user":{"displayName":"Simon Aubé","userId":"10741788869764524869"}}},"execution_count":46,"outputs":[]},{"cell_type":"code","source":["rates_corrected"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":540},"id":"APhwjAP15XVC","executionInfo":{"status":"ok","timestamp":1671477882018,"user_tz":300,"elapsed":5,"user":{"displayName":"Simon Aubé","userId":"10741788869764524869"}},"outputId":"2747f596-d71b-49a3-8c6d-b5bc27c5c52e"},"execution_count":47,"outputs":[{"output_type":"execute_result","data":{"text/plain":["          ORF   Gene        lm        lp      wRPF     wmRNA    bmEser  \\\n","0     YAL001C   TFC3  3.553033  3.064832  1.258158  1.577492  1.125383   \n","1     YAL002W   VPS8  3.582631  3.105510  0.758155  1.345766  1.146577   \n","2     YAL003W   EFB1  2.994317  2.315970  3.849116  3.664817  2.307282   \n","3     YAL007C   ERP2  2.811575  2.334454  2.321122  2.327195  1.701383   \n","4     YAL008W  FUN14  2.775974  2.298853  1.980231  1.698622  0.790021   \n","...       ...    ...       ...       ...       ...       ...       ...   \n","4435  YPR194C   OPT2  3.420616  2.943495 -0.236572  0.842609 -0.399593   \n","4436  YPR198W   SGE1  3.212720  2.735599  1.472025  1.521922  1.158760   \n","4437  YPR199C   ARR1  2.946943  2.469822  1.138934  1.592954  1.449894   \n","4438  YPR200C   ARR2  2.594393  2.117271  0.387390  1.169086  0.400387   \n","4439  YPR201W   ARR3  3.084576  2.607455  0.305351  0.501059 -0.077695   \n","\n","        amEser        bm         m  ...  isEssential  hasTATA   YEPDFit  \\\n","0     0.739048  1.127728  0.423025  ...         True     True  0.040325   \n","1     0.892427  0.896002  0.191299  ...        False    False  0.020725   \n","2     0.194135  3.215053  2.510350  ...         True    False  0.420089   \n","3     0.193896  1.877432  1.172729  ...        False    False -0.026295   \n","4     0.452387  1.248859  0.544156  ...        False    False -0.018375   \n","...        ...       ...       ...  ...          ...      ...       ...   \n","4435  0.362062  0.392846 -0.311857  ...        False     True -0.009625   \n","4436  0.582951  1.072159  0.367456  ...        False    False -0.009200   \n","4437  0.762928  1.143190  0.438487  ...        False    False -0.016960   \n","4438  0.888198  0.719323  0.014620  ...        False     True -0.010200   \n","4439  0.738003  0.051296 -0.653407  ...        False     True -0.008790   \n","\n","          pEst     Duplication     m_est   bm_calc   bp_calc pEst_calc  \\\n","0     2.496036             SSD  0.423053  1.130623  2.956097  3.252044   \n","1     1.955355  SSD-successive  0.191327  0.898897  2.687819  2.752041   \n","2     5.835856               S  2.510378  3.217948  3.459730  5.843003   \n","3     4.289378         SSD_WGD  1.172756  1.880326  3.269357  4.315008   \n","4     3.984088               S  0.544183  1.251753  3.557038  3.974117   \n","...        ...             ...       ...       ...       ...       ...   \n","4435  1.122644             SSD -0.311830  0.395740  2.196249  1.757314   \n","4436  3.039136  SSD-successive  0.367483  1.075053  3.225533  3.465911   \n","4437  2.971822  SSD-successive  0.438514  1.146085  2.821411  3.132820   \n","4438  2.572829             SSD  0.014647  0.722217  2.493734  2.381276   \n","4439  2.000607               S -0.653380  0.054190  3.079722  2.299238   \n","\n","       bm_Eser  \n","0     1.162073  \n","1     1.083726  \n","2     2.704485  \n","3     1.366624  \n","4     0.996543  \n","...        ...  \n","4435  0.050205  \n","4436  0.950407  \n","4437  1.201415  \n","4438  0.902818  \n","4439  0.084596  \n","\n","[4440 rows x 24 columns]"],"text/html":["\n","  <div id=\"df-2800800f-46ed-4882-a794-7102893dff8f\">\n","    <div class=\"colab-df-container\">\n","      <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>ORF</th>\n","      <th>Gene</th>\n","      <th>lm</th>\n","      <th>lp</th>\n","      <th>wRPF</th>\n","      <th>wmRNA</th>\n","      <th>bmEser</th>\n","      <th>amEser</th>\n","      <th>bm</th>\n","      <th>m</th>\n","      <th>...</th>\n","      <th>isEssential</th>\n","      <th>hasTATA</th>\n","      <th>YEPDFit</th>\n","      <th>pEst</th>\n","      <th>Duplication</th>\n","      <th>m_est</th>\n","      <th>bm_calc</th>\n","      <th>bp_calc</th>\n","      <th>pEst_calc</th>\n","      <th>bm_Eser</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>YAL001C</td>\n","      <td>TFC3</td>\n","      <td>3.553033</td>\n","      <td>3.064832</td>\n","      <td>1.258158</td>\n","      <td>1.577492</td>\n","      <td>1.125383</td>\n","      <td>0.739048</td>\n","      <td>1.127728</td>\n","      <td>0.423025</td>\n","      <td>...</td>\n","      <td>True</td>\n","      <td>True</td>\n","      <td>0.040325</td>\n","      <td>2.496036</td>\n","      <td>SSD</td>\n","      <td>0.423053</td>\n","      <td>1.130623</td>\n","      <td>2.956097</td>\n","      <td>3.252044</td>\n","      <td>1.162073</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>YAL002W</td>\n","      <td>VPS8</td>\n","      <td>3.582631</td>\n","      <td>3.105510</td>\n","      <td>0.758155</td>\n","      <td>1.345766</td>\n","      <td>1.146577</td>\n","      <td>0.892427</td>\n","      <td>0.896002</td>\n","      <td>0.191299</td>\n","      <td>...</td>\n","      <td>False</td>\n","      <td>False</td>\n","      <td>0.020725</td>\n","      <td>1.955355</td>\n","      <td>SSD-successive</td>\n","      <td>0.191327</td>\n","      <td>0.898897</td>\n","      <td>2.687819</td>\n","      <td>2.752041</td>\n","      <td>1.083726</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>YAL003W</td>\n","      <td>EFB1</td>\n","      <td>2.994317</td>\n","      <td>2.315970</td>\n","      <td>3.849116</td>\n","      <td>3.664817</td>\n","      <td>2.307282</td>\n","      <td>0.194135</td>\n","      <td>3.215053</td>\n","      <td>2.510350</td>\n","      <td>...</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>0.420089</td>\n","      <td>5.835856</td>\n","      <td>S</td>\n","      <td>2.510378</td>\n","      <td>3.217948</td>\n","      <td>3.459730</td>\n","      <td>5.843003</td>\n","      <td>2.704485</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>YAL007C</td>\n","      <td>ERP2</td>\n","      <td>2.811575</td>\n","      <td>2.334454</td>\n","      <td>2.321122</td>\n","      <td>2.327195</td>\n","      <td>1.701383</td>\n","      <td>0.193896</td>\n","      <td>1.877432</td>\n","      <td>1.172729</td>\n","      <td>...</td>\n","      <td>False</td>\n","      <td>False</td>\n","      <td>-0.026295</td>\n","      <td>4.289378</td>\n","      <td>SSD_WGD</td>\n","      <td>1.172756</td>\n","      <td>1.880326</td>\n","      <td>3.269357</td>\n","      <td>4.315008</td>\n","      <td>1.366624</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>YAL008W</td>\n","      <td>FUN14</td>\n","      <td>2.775974</td>\n","      <td>2.298853</td>\n","      <td>1.980231</td>\n","      <td>1.698622</td>\n","      <td>0.790021</td>\n","      <td>0.452387</td>\n","      <td>1.248859</td>\n","      <td>0.544156</td>\n","      <td>...</td>\n","      <td>False</td>\n","      <td>False</td>\n","      <td>-0.018375</td>\n","      <td>3.984088</td>\n","      <td>S</td>\n","      <td>0.544183</td>\n","      <td>1.251753</td>\n","      <td>3.557038</td>\n","      <td>3.974117</td>\n","      <td>0.996543</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>4435</th>\n","      <td>YPR194C</td>\n","      <td>OPT2</td>\n","      <td>3.420616</td>\n","      <td>2.943495</td>\n","      <td>-0.236572</td>\n","      <td>0.842609</td>\n","      <td>-0.399593</td>\n","      <td>0.362062</td>\n","      <td>0.392846</td>\n","      <td>-0.311857</td>\n","      <td>...</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>-0.009625</td>\n","      <td>1.122644</td>\n","      <td>SSD</td>\n","      <td>-0.311830</td>\n","      <td>0.395740</td>\n","      <td>2.196249</td>\n","      <td>1.757314</td>\n","      <td>0.050205</td>\n","    </tr>\n","    <tr>\n","      <th>4436</th>\n","      <td>YPR198W</td>\n","      <td>SGE1</td>\n","      <td>3.212720</td>\n","      <td>2.735599</td>\n","      <td>1.472025</td>\n","      <td>1.521922</td>\n","      <td>1.158760</td>\n","      <td>0.582951</td>\n","      <td>1.072159</td>\n","      <td>0.367456</td>\n","      <td>...</td>\n","      <td>False</td>\n","      <td>False</td>\n","      <td>-0.009200</td>\n","      <td>3.039136</td>\n","      <td>SSD-successive</td>\n","      <td>0.367483</td>\n","      <td>1.075053</td>\n","      <td>3.225533</td>\n","      <td>3.465911</td>\n","      <td>0.950407</td>\n","    </tr>\n","    <tr>\n","      <th>4437</th>\n","      <td>YPR199C</td>\n","      <td>ARR1</td>\n","      <td>2.946943</td>\n","      <td>2.469822</td>\n","      <td>1.138934</td>\n","      <td>1.592954</td>\n","      <td>1.449894</td>\n","      <td>0.762928</td>\n","      <td>1.143190</td>\n","      <td>0.438487</td>\n","      <td>...</td>\n","      <td>False</td>\n","      <td>False</td>\n","      <td>-0.016960</td>\n","      <td>2.971822</td>\n","      <td>SSD-successive</td>\n","      <td>0.438514</td>\n","      <td>1.146085</td>\n","      <td>2.821411</td>\n","      <td>3.132820</td>\n","      <td>1.201415</td>\n","    </tr>\n","    <tr>\n","      <th>4438</th>\n","      <td>YPR200C</td>\n","      <td>ARR2</td>\n","      <td>2.594393</td>\n","      <td>2.117271</td>\n","      <td>0.387390</td>\n","      <td>1.169086</td>\n","      <td>0.400387</td>\n","      <td>0.888198</td>\n","      <td>0.719323</td>\n","      <td>0.014620</td>\n","      <td>...</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>-0.010200</td>\n","      <td>2.572829</td>\n","      <td>SSD</td>\n","      <td>0.014647</td>\n","      <td>0.722217</td>\n","      <td>2.493734</td>\n","      <td>2.381276</td>\n","      <td>0.902818</td>\n","    </tr>\n","    <tr>\n","      <th>4439</th>\n","      <td>YPR201W</td>\n","      <td>ARR3</td>\n","      <td>3.084576</td>\n","      <td>2.607455</td>\n","      <td>0.305351</td>\n","      <td>0.501059</td>\n","      <td>-0.077695</td>\n","      <td>0.738003</td>\n","      <td>0.051296</td>\n","      <td>-0.653407</td>\n","      <td>...</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>-0.008790</td>\n","      <td>2.000607</td>\n","      <td>S</td>\n","      <td>-0.653380</td>\n","      <td>0.054190</td>\n","      <td>3.079722</td>\n","      <td>2.299238</td>\n","      <td>0.084596</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>4440 rows × 24 columns</p>\n","</div>\n","      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-2800800f-46ed-4882-a794-7102893dff8f')\"\n","              title=\"Convert this dataframe to an interactive table.\"\n","              style=\"display:none;\">\n","        \n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","       width=\"24px\">\n","    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n","    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n","  </svg>\n","      </button>\n","      \n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      flex-wrap:wrap;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","      <script>\n","        const buttonEl =\n","          document.querySelector('#df-2800800f-46ed-4882-a794-7102893dff8f button.colab-df-convert');\n","        buttonEl.style.display =\n","          google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","        async function convertToInteractive(key) {\n","          const element = document.querySelector('#df-2800800f-46ed-4882-a794-7102893dff8f');\n","          const dataTable =\n","            await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                     [key], {});\n","          if (!dataTable) return;\n","\n","          const docLinkHtml = 'Like what you see? Visit the ' +\n","            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","            + ' to learn more about interactive tables.';\n","          element.innerHTML = '';\n","          dataTable['output_type'] = 'display_data';\n","          await google.colab.output.renderOutput(dataTable, element);\n","          const docLink = document.createElement('div');\n","          docLink.innerHTML = docLinkHtml;\n","          element.appendChild(docLink);\n","        }\n","      </script>\n","    </div>\n","  </div>\n","  "]},"metadata":{},"execution_count":47}]},{"cell_type":"markdown","metadata":{"id":"4T1TNA5BjePG"},"source":["This is the final rates_Hausser dataframe which was used to initialize the simulations. It is exported:"]},{"cell_type":"code","metadata":{"id":"zGXAbCDkjoCg","executionInfo":{"status":"ok","timestamp":1671477882318,"user_tz":300,"elapsed":304,"user":{"displayName":"Simon Aubé","userId":"10741788869764524869"}}},"source":["rates_corrected.to_csv(path_or_buf='/content/drive/My Drive/Redaction_SA/Figures_2021_06/Data_sim_ready/rates_Hausser.csv', index=False, header=True)"],"execution_count":48,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"bLhBkngWksAw"},"source":["## 4) Reconstitution of paralog couples and calculation of relative divergence"]},{"cell_type":"markdown","metadata":{"id":"nq0yNQ8IpBeP"},"source":["The P1-P2 associations are taken from Table S9 of Supplementary File 2 from Marchant et al. (2019), which compiles both SSD- and WGD-derived paralog pairs (as well as other types of paralogs such as multi-SSD which are ignored in the current work). \n"]},{"cell_type":"code","metadata":{"id":"S0q5sXOCoRQS","executionInfo":{"status":"ok","timestamp":1671477882541,"user_tz":300,"elapsed":224,"user":{"displayName":"Simon Aubé","userId":"10741788869764524869"}}},"source":["dupli_pairs = pd.read_csv('/content/drive/My Drive/Redaction_SA/Figures_2021_06/Data_init/TableS9_Marchant_2019.csv', sep=';')\n","dupli_pairs = dupli_pairs.dropna(subset=['Duplication', 'P1', 'P2']).reset_index(drop=True)"],"execution_count":49,"outputs":[]},{"cell_type":"code","metadata":{"id":"_hohIJCasWsN","executionInfo":{"status":"ok","timestamp":1671477882541,"user_tz":300,"elapsed":3,"user":{"displayName":"Simon Aubé","userId":"10741788869764524869"}}},"source":["# A new dataframe is defined\n","rates_couples = pd.DataFrame(columns=['Duplication', 'ORF_P1', 'ORF_P2'])\n","\n","# After merging, the dataframe will have all those columns:\n","# Duplication: Duplication type from Marchant et al.\n","# ORF_P1: Systematic name paralog 1\n","# Gene_P1: Standard name paralog 1\n","# ORF_P2: Systematic name paralog 2\n","# Gene_P2: Standard name paralog 2\n","# Couple: P1-P1 (standard names)\n","# bm_P1: Log10 transcription rate of P1, in mRNAs/h (Hausser et al.)\n","# bm_P2: Log10 transcription rate of P2, in mRNAs/h (Hausser et al.)\n","# m_P1: Log10 mRNA abundance for P1, in mRNAs (Hausser et al.)\n","# m_P2: Log10 mRNA abundance for P2, in mRNAs (Hausser et al.)\n","# amEser_P1: Log10 decay rate of mRNA P1, in h^-1 (Eser et al.)\n","# amEser_P2: Log10 decay rate of mRNA P2, in h^-1 (Eser et al.)\n","# bm_Eser_P1: Log10 transcription rate of P1 recalculated to account for mRNA\n","#             decay measurements (Eser et al.), in mRNAs/h (Hausser et al.)\n","# bm_Eser_P2: Log10 transcription rate of P1 recalculated to account for mRNA\n","#             decay measurements (Eser et al.), in mRNAs/h (Hausser et al.)\n","# bp_P1: Log10 Translation rate of P1, in proteins per mRNA per hour (Hausser et al.)\n","# bp_calc_P1: Recalculated translation rate of P1\n","# bp_P2: Log10 Translation rate of P2, in proteins per mRNA per hour (Hausser et al.)\n","# bp_calc_P2: Recalculated translation rate of P2\n","# pEst_P1: Log10 Estimated protein abundance for P1 (proteins / cell)\n","# pEst_calc_P1: Log10 Estimated protein abundance for P1, recalculated using the\n","#               recalculated translation rate (proteins / cell)\n","# pEst_P2: Log10 Estimated protein abundance for P2 (proteins per cell)\n","# pEst_calc_P2: Log10 Estimated protein abundance for P2, recalculated using the\n","#               recalculated translation rate (proteins / cell)\n","# apBelle_P1: Log10 protein decay rate for P1, in h^-1 (Belle et al.)\n","# apBelle_P2: Log10 protein decay rate for P2, in h^-1 (Belle et al.)\n","# cv_P1: Coefficient of variation for P1 (Newman et al., 2006)\n","# cv_P2: Coefficient of variation for P2 (Newman et al., 2006)\n","\n","rates_couples['Duplication'] = dupli_pairs['Duplication']\n","rates_couples['ORF_P1'] = dupli_pairs['P1']\n","rates_couples['ORF_P2'] = dupli_pairs['P2']"],"execution_count":50,"outputs":[]},{"cell_type":"code","metadata":{"id":"aKQYl14-0Qte","executionInfo":{"status":"ok","timestamp":1671477882541,"user_tz":300,"elapsed":3,"user":{"displayName":"Simon Aubé","userId":"10741788869764524869"}}},"source":["# New versions of the rates_corrected dataframe are prepared to be merged with\n","# the rates_couples df just created\n","\n","rates_only = rates_corrected[['ORF', 'Gene', 'bm', 'm', 'amEser', 'bm_Eser',\n","                              'bp', 'cv', 'pEst', 'apExp', 'Duplication',\n","                              'bp_calc', 'pEst_calc']].copy()\n","\n","rates_P1 = rates_only.copy()\n","rates_P1 = rates_P1.rename(columns={'ORF': 'ORF_P1', 'Gene': 'Gene_P1', \n","                                    'bm': 'bm_P1', 'm': 'm_P1', \n","                                    'amEser': 'amEser_P1', \n","                                    'bm_Eser': 'bm_Eser_P1', 'bp': 'bp_P1', \n","                                    'cv': 'cv_P1', 'pEst': 'pEst_P1',\n","                                    'apExp': 'apBelle_P1', \n","                                    'bp_calc': 'bp_calc_P1',\n","                                    'pEst_calc': 'pEst_calc_P1'})\n","\n","rates_P2 = rates_only.copy()\n","rates_P2 = rates_P2.rename(columns={'ORF': 'ORF_P2', 'Gene': 'Gene_P2', \n","                                    'bm': 'bm_P2', 'm': 'm_P2',\n","                                    'amEser': 'amEser_P2',\n","                                    'bm_Eser': 'bm_Eser_P2', 'bp': 'bp_P2',\n","                                    'cv': 'cv_P2', 'pEst': 'pEst_P2',\n","                                    'apExp': 'apBelle_P2',\n","                                    'bp_calc': 'bp_calc_P2',\n","                                    'pEst_calc': 'pEst_calc_P2'})"],"execution_count":51,"outputs":[]},{"cell_type":"code","metadata":{"id":"KFWUpWLe1whp","executionInfo":{"status":"ok","timestamp":1671477882542,"user_tz":300,"elapsed":4,"user":{"displayName":"Simon Aubé","userId":"10741788869764524869"}}},"source":["# The dataframe is then filled\n","rates_couples = pd.merge(rates_couples, rates_P1, how='left',\n","                         on=['Duplication', 'ORF_P1'])\n","\n","rates_couples = pd.merge(rates_couples, rates_P2, how='left',\n","                         on=['Duplication', 'ORF_P2'])"],"execution_count":52,"outputs":[]},{"cell_type":"code","metadata":{"id":"CQSSIYxWcaWN","executionInfo":{"status":"ok","timestamp":1671477882542,"user_tz":300,"elapsed":4,"user":{"displayName":"Simon Aubé","userId":"10741788869764524869"}}},"source":["# Keeping only pairs where both paralogs were included in the dataset\n","rates_couples = rates_couples.dropna(subset=['bm_P1', 'bm_P2', 'bp_P1', 'bp_P2']).reset_index(drop=True)"],"execution_count":53,"outputs":[]},{"cell_type":"code","metadata":{"id":"nEgUx58-rsSO","executionInfo":{"status":"ok","timestamp":1671477882542,"user_tz":300,"elapsed":3,"user":{"displayName":"Simon Aubé","userId":"10741788869764524869"}}},"source":["# Rearranging the columns in the desired order:\n","rates_couples = rates_couples[['Duplication', 'ORF_P1', 'Gene_P1', 'ORF_P2', \n","                               'Gene_P2', 'bm_P1', 'bm_P2', 'm_P1', 'm_P2',\n","                               'amEser_P1', 'amEser_P2', 'bm_Eser_P1',\n","                               'bm_Eser_P2', 'bp_P1', 'bp_calc_P1', 'bp_P2',\n","                               'bp_calc_P2', 'pEst_P1', 'pEst_calc_P1', \n","                               'pEst_P2', 'pEst_calc_P2', 'apBelle_P1',\n","                               'apBelle_P2', 'cv_P1', 'cv_P2']]"],"execution_count":54,"outputs":[]},{"cell_type":"code","metadata":{"id":"ywWhZ1-xvQWH","executionInfo":{"status":"ok","timestamp":1671477882678,"user_tz":300,"elapsed":139,"user":{"displayName":"Simon Aubé","userId":"10741788869764524869"}}},"source":["# Definition of a function to compute log2 fold-changes of relative divergence\n","\n","def fold_change(prop_P1, prop_P2, data, name):\n","  \"\"\"Function to calculate log2 fold-change for a property between two\n","     duplicates. The two properties are provided as dataframe columns.\n","     The last argument is used to specify the name of the new column \n","     inserted in the data dataframe.\"\"\"\n","\n","  df = data.copy()\n","  df[name] = np.NaN\n","\n","  for row in range(df.shape[0]):\n","\n","    value_P1 = df.at[row, f'{prop_P1}']\n","    value_P2 = df.at[row, f'{prop_P2}']\n","\n","    if value_P1 >= value_P2:\n","\n","      df.at[row, name] = (10**value_P1)/(10**value_P2)\n","\n","    elif value_P1 < value_P2:\n","\n","      df.at[row, name] = (10**value_P2)/(10**value_P1)\n","\n","  df[name] = np.log2(df[name])\n","\n","  return df"],"execution_count":55,"outputs":[]},{"cell_type":"code","metadata":{"id":"LBNRt4DAv5co","executionInfo":{"status":"ok","timestamp":1671477882679,"user_tz":300,"elapsed":3,"user":{"displayName":"Simon Aubé","userId":"10741788869764524869"}}},"source":["# Calculation of the log2 fold-changes of divergence\n","rates_logfold = fold_change('bm_P1', 'bm_P2', rates_couples, 'bm_fold_original')\n","rates_logfold = fold_change('amEser_P1', 'amEser_P2', rates_logfold, 'amEser_fold')\n","rates_logfold = fold_change('bm_Eser_P1', 'bm_Eser_P2', rates_logfold, 'bm_fold_Eser')\n","rates_logfold = fold_change('bp_P1', 'bp_P2', rates_logfold, 'bp_fold_original')\n","rates_logfold = fold_change('pEst_P1', 'pEst_P2', rates_logfold, 'pEst_fold_original')\n","rates_logfold = fold_change('apBelle_P1', 'apBelle_P2', rates_logfold, 'apBelle_fold')\n","\n","rates_logfold = fold_change('bp_calc_P1', 'bp_calc_P2', rates_logfold, 'bp_fold_calc')\n","rates_logfold = fold_change('pEst_calc_P1', 'pEst_calc_P2', rates_logfold, 'pEst_fold_calc')"],"execution_count":56,"outputs":[]},{"cell_type":"code","metadata":{"id":"WwdOtbaJyYNQ","executionInfo":{"status":"ok","timestamp":1671477882679,"user_tz":300,"elapsed":3,"user":{"displayName":"Simon Aubé","userId":"10741788869764524869"}}},"source":["# Calculation of normalized differences\n","rates_div_rel = rates_logfold.copy()\n","\n","rates_div_rel['bm_diff'] = np.abs(10**rates_div_rel['bm_P2'] - 10**rates_div_rel['bm_P1']) / (10**rates_div_rel['bm_P2'] + 10**rates_div_rel['bm_P1'])\n","rates_div_rel['bm_Eser_diff'] = np.abs(10**rates_div_rel['bm_Eser_P2'] - 10**rates_div_rel['bm_Eser_P1']) / (10**rates_div_rel['bm_Eser_P2'] + 10**rates_div_rel['bm_Eser_P1'])\n","rates_div_rel['amEser_diff'] = np.abs(10**rates_div_rel['amEser_P2'] - 10**rates_div_rel['amEser_P1']) / (10**rates_div_rel['amEser_P2'] + 10**rates_div_rel['amEser_P1'])\n","rates_div_rel['bp_diff'] = np.abs(10**rates_div_rel['bp_P2'] - 10**rates_div_rel['bp_P1']) / (10**rates_div_rel['bp_P2'] + 10**rates_div_rel['bp_P1'])\n","rates_div_rel['pEst_diff'] = np.abs(10**rates_div_rel['pEst_P2'] - 10**rates_div_rel['pEst_P1']) / (10**rates_div_rel['pEst_P2'] + 10**rates_div_rel['pEst_P1'])\n","rates_div_rel['apBelle_diff'] = np.abs(10**rates_div_rel['apBelle_P2'] - 10**rates_div_rel['apBelle_P1']) / (10**rates_div_rel['apBelle_P2'] + 10**rates_div_rel['apBelle_P1'])\n","rates_div_rel['bp_calc_diff'] = np.abs(10**rates_div_rel['bp_calc_P2'] - 10**rates_div_rel['bp_calc_P1']) / (10**rates_div_rel['bp_calc_P2'] + 10**rates_div_rel['bp_calc_P1'])\n","rates_div_rel['pEst_calc_diff'] = np.abs(10**rates_div_rel['pEst_calc_P2'] - 10**rates_div_rel['pEst_calc_P1']) / (10**rates_div_rel['pEst_calc_P2'] + 10**rates_div_rel['pEst_calc_P1'])"],"execution_count":57,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"HTMiEtAt0y40"},"source":["This is the final dataframe with log2 fold-changes and normalized differences that is used to generate figure 1 as well as to initialize the simulations (and compare the simulated and empirical data at the end). It is exported as csv:"]},{"cell_type":"code","metadata":{"id":"3YWb3Z7A1UxX","executionInfo":{"status":"ok","timestamp":1671477883060,"user_tz":300,"elapsed":384,"user":{"displayName":"Simon Aubé","userId":"10741788869764524869"}}},"source":["rates_div_rel.to_csv(path_or_buf='/content/drive/My Drive/Redaction_SA/Figures_2021_06/Data_sim_ready/couples_divergence.csv', index=False, header=True)"],"execution_count":58,"outputs":[]}]}