{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"Data_preparation.ipynb","provenance":[],"authorship_tag":"ABX9TyOjebywv8r+TG/favqvCinp"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","metadata":{"id":"kqdNELCCOMPy"},"source":["#**Script to obtain the data files that are used to initialize the simulations of transcriptional and translational divergence of yeast duplicated genes**"]},{"cell_type":"code","metadata":{"id":"nlxHPg2AO2tt"},"source":["import numpy as np\n","import pandas as pd\n","from google.colab import files\n","import io\n","import scipy.stats as stats"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"XtWuZ-afO_GO","executionInfo":{"status":"ok","timestamp":1645459422978,"user_tz":300,"elapsed":19306,"user":{"displayName":"Simon Aubé","photoUrl":"https://lh3.googleusercontent.com/a/default-user=s64","userId":"10741788869764524869"}},"outputId":"99358d84-24c8-4b6e-96af-a98d448034d7"},"source":["# Mounting my google drive\n","\n","from google.colab import drive\n","drive.mount('/content/drive')"],"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Mounted at /content/drive\n"]}]},{"cell_type":"markdown","metadata":{"id":"PJ72tXC9OmxI"},"source":["Transcription and translation rates from Hausser et al. (2019) are first imported."]},{"cell_type":"code","metadata":{"id":"5g2pbW6jOCyo"},"source":["rates_yeast = pd.read_csv('/content/drive/My Drive/Redaction_SA/Figures_2021_06/Data_init/yeastRates.csv', sep=';', decimal=',')\n","rates_yeast = rates_yeast.rename(columns={'Unnamed: 0': 'Gene'})"],"execution_count":null,"outputs":[]},{"cell_type":"code","source":["rates_yeast"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":424},"id":"fAh7G8TaBmJT","executionInfo":{"status":"ok","timestamp":1645459438307,"user_tz":300,"elapsed":434,"user":{"displayName":"Simon Aubé","photoUrl":"https://lh3.googleusercontent.com/a/default-user=s64","userId":"10741788869764524869"}},"outputId":"efb3c804-3ecb-4c1d-a4a7-bcdc6f06bbfe"},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/html":["\n","  <div id=\"df-0b55bce0-ce63-4322-8ea7-6aa2370889c2\">\n","    <div class=\"colab-df-container\">\n","      <div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Gene</th>\n","      <th>lm</th>\n","      <th>lp</th>\n","      <th>wRPF</th>\n","      <th>wmRNA</th>\n","      <th>bmEser</th>\n","      <th>amEser</th>\n","      <th>bm</th>\n","      <th>m</th>\n","      <th>bp</th>\n","      <th>cv</th>\n","      <th>apExp</th>\n","      <th>am</th>\n","      <th>isEssential</th>\n","      <th>hasTATA</th>\n","      <th>YEPDFit</th>\n","      <th>pEst</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>TFC3</td>\n","      <td>3.553033</td>\n","      <td>3.064832</td>\n","      <td>1.258158</td>\n","      <td>1.577492</td>\n","      <td>1.125383</td>\n","      <td>0.704432</td>\n","      <td>1.127728</td>\n","      <td>0.423025</td>\n","      <td>2.201503</td>\n","      <td>NaN</td>\n","      <td>-1.057717</td>\n","      <td>0.704302</td>\n","      <td>True</td>\n","      <td>True</td>\n","      <td>0.040325</td>\n","      <td>2.496036</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>VPS8</td>\n","      <td>3.582631</td>\n","      <td>3.105510</td>\n","      <td>0.758155</td>\n","      <td>1.345766</td>\n","      <td>1.146577</td>\n","      <td>0.868402</td>\n","      <td>0.896002</td>\n","      <td>0.191299</td>\n","      <td>1.892547</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>0.704302</td>\n","      <td>False</td>\n","      <td>False</td>\n","      <td>0.020725</td>\n","      <td>1.955355</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>EFB1</td>\n","      <td>2.994317</td>\n","      <td>2.315970</td>\n","      <td>3.849116</td>\n","      <td>3.664817</td>\n","      <td>2.307282</td>\n","      <td>0.058252</td>\n","      <td>3.215053</td>\n","      <td>2.510350</td>\n","      <td>3.453998</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>0.704302</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>0.420089</td>\n","      <td>5.835856</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>ERP2</td>\n","      <td>2.811575</td>\n","      <td>2.334454</td>\n","      <td>2.321122</td>\n","      <td>2.327195</td>\n","      <td>1.701383</td>\n","      <td>0.057926</td>\n","      <td>1.877432</td>\n","      <td>1.172729</td>\n","      <td>3.245141</td>\n","      <td>-0.744004</td>\n","      <td>-0.071219</td>\n","      <td>0.704302</td>\n","      <td>False</td>\n","      <td>False</td>\n","      <td>-0.026295</td>\n","      <td>4.289378</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>FUN14</td>\n","      <td>2.775974</td>\n","      <td>2.298853</td>\n","      <td>1.980231</td>\n","      <td>1.698622</td>\n","      <td>0.790021</td>\n","      <td>0.382707</td>\n","      <td>1.248859</td>\n","      <td>0.544156</td>\n","      <td>3.568424</td>\n","      <td>NaN</td>\n","      <td>0.204003</td>\n","      <td>0.704302</td>\n","      <td>False</td>\n","      <td>False</td>\n","      <td>-0.018375</td>\n","      <td>3.984088</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>4435</th>\n","      <td>OPT2</td>\n","      <td>3.420616</td>\n","      <td>2.943495</td>\n","      <td>-0.236572</td>\n","      <td>0.842609</td>\n","      <td>-0.399593</td>\n","      <td>0.274546</td>\n","      <td>0.392846</td>\n","      <td>-0.311857</td>\n","      <td>1.562993</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>0.704302</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>-0.009625</td>\n","      <td>1.122644</td>\n","    </tr>\n","    <tr>\n","      <th>4436</th>\n","      <td>SGE1</td>\n","      <td>3.212720</td>\n","      <td>2.735599</td>\n","      <td>1.472025</td>\n","      <td>1.521922</td>\n","      <td>1.158760</td>\n","      <td>0.532465</td>\n","      <td>1.072159</td>\n","      <td>0.367456</td>\n","      <td>2.800172</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>0.704302</td>\n","      <td>False</td>\n","      <td>False</td>\n","      <td>-0.009200</td>\n","      <td>3.039136</td>\n","    </tr>\n","    <tr>\n","      <th>4437</th>\n","      <td>ARR1</td>\n","      <td>2.946943</td>\n","      <td>2.469822</td>\n","      <td>1.138934</td>\n","      <td>1.592954</td>\n","      <td>1.449894</td>\n","      <td>0.730236</td>\n","      <td>1.143190</td>\n","      <td>0.438487</td>\n","      <td>2.661827</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>0.704302</td>\n","      <td>False</td>\n","      <td>False</td>\n","      <td>-0.016960</td>\n","      <td>2.971822</td>\n","    </tr>\n","    <tr>\n","      <th>4438</th>\n","      <td>ARR2</td>\n","      <td>2.594393</td>\n","      <td>2.117271</td>\n","      <td>0.387390</td>\n","      <td>1.169086</td>\n","      <td>0.400387</td>\n","      <td>0.863932</td>\n","      <td>0.719323</td>\n","      <td>0.014620</td>\n","      <td>2.686701</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>0.704302</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>-0.010200</td>\n","      <td>2.572829</td>\n","    </tr>\n","    <tr>\n","      <th>4439</th>\n","      <td>ARR3</td>\n","      <td>3.084576</td>\n","      <td>2.607455</td>\n","      <td>0.305351</td>\n","      <td>0.501059</td>\n","      <td>-0.077695</td>\n","      <td>0.703301</td>\n","      <td>0.051296</td>\n","      <td>-0.653407</td>\n","      <td>2.782506</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>0.704302</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>-0.008790</td>\n","      <td>2.000607</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>4440 rows × 17 columns</p>\n","</div>\n","      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-0b55bce0-ce63-4322-8ea7-6aa2370889c2')\"\n","              title=\"Convert this dataframe to an interactive table.\"\n","              style=\"display:none;\">\n","        \n","  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n","       width=\"24px\">\n","    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n","    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n","  </svg>\n","      </button>\n","      \n","  <style>\n","    .colab-df-container {\n","      display:flex;\n","      flex-wrap:wrap;\n","      gap: 12px;\n","    }\n","\n","    .colab-df-convert {\n","      background-color: #E8F0FE;\n","      border: none;\n","      border-radius: 50%;\n","      cursor: pointer;\n","      display: none;\n","      fill: #1967D2;\n","      height: 32px;\n","      padding: 0 0 0 0;\n","      width: 32px;\n","    }\n","\n","    .colab-df-convert:hover {\n","      background-color: #E2EBFA;\n","      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n","      fill: #174EA6;\n","    }\n","\n","    [theme=dark] .colab-df-convert {\n","      background-color: #3B4455;\n","      fill: #D2E3FC;\n","    }\n","\n","    [theme=dark] .colab-df-convert:hover {\n","      background-color: #434B5C;\n","      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n","      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n","      fill: #FFFFFF;\n","    }\n","  </style>\n","\n","      <script>\n","        const buttonEl =\n","          document.querySelector('#df-0b55bce0-ce63-4322-8ea7-6aa2370889c2 button.colab-df-convert');\n","        buttonEl.style.display =\n","          google.colab.kernel.accessAllowed ? 'block' : 'none';\n","\n","        async function convertToInteractive(key) {\n","          const element = document.querySelector('#df-0b55bce0-ce63-4322-8ea7-6aa2370889c2');\n","          const dataTable =\n","            await google.colab.kernel.invokeFunction('convertToInteractive',\n","                                                     [key], {});\n","          if (!dataTable) return;\n","\n","          const docLinkHtml = 'Like what you see? Visit the ' +\n","            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n","            + ' to learn more about interactive tables.';\n","          element.innerHTML = '';\n","          dataTable['output_type'] = 'display_data';\n","          await google.colab.output.renderOutput(dataTable, element);\n","          const docLink = document.createElement('div');\n","          docLink.innerHTML = docLinkHtml;\n","          element.appendChild(docLink);\n","        }\n","      </script>\n","    </div>\n","  </div>\n","  "],"text/plain":["       Gene        lm        lp  ...  hasTATA   YEPDFit      pEst\n","0      TFC3  3.553033  3.064832  ...     True  0.040325  2.496036\n","1      VPS8  3.582631  3.105510  ...    False  0.020725  1.955355\n","2      EFB1  2.994317  2.315970  ...    False  0.420089  5.835856\n","3      ERP2  2.811575  2.334454  ...    False -0.026295  4.289378\n","4     FUN14  2.775974  2.298853  ...    False -0.018375  3.984088\n","...     ...       ...       ...  ...      ...       ...       ...\n","4435   OPT2  3.420616  2.943495  ...     True -0.009625  1.122644\n","4436   SGE1  3.212720  2.735599  ...    False -0.009200  3.039136\n","4437   ARR1  2.946943  2.469822  ...    False -0.016960  2.971822\n","4438   ARR2  2.594393  2.117271  ...     True -0.010200  2.572829\n","4439   ARR3  3.084576  2.607455  ...     True -0.008790  2.000607\n","\n","[4440 rows x 17 columns]"]},"metadata":{},"execution_count":4}]},{"cell_type":"markdown","metadata":{"id":"ayF5mcI0WKrN"},"source":["## 1) Addition of systematic names"]},{"cell_type":"markdown","metadata":{"id":"-5cPWGM1bT69"},"source":["The file saccharomyces_cerevisiae.20200611.gff has been downloaded from Saccharomyces Genome Database (SGD). It contains all genomic features in the S288C reference genome, as of 2020-06-11, as well as the complete genome sequence. It is used to obtain all the systematic - standard name associations."]},{"cell_type":"code","metadata":{"id":"0md2sFErf4eq"},"source":["features_SGD = pd.read_table('/content/drive/My Drive/Redaction_SA/Figures_2021_06/Data_init/saccharomyces_cerevisiae.20200611.gff', \n","                             skiprows=20, header=None, index_col=False, \n","                             names=['seqname', 'source', 'feature', 'start', \n","                                    'end', 'score', 'strand', 'frame', \n","                                    'attribute'], nrows=23067)\n","# The genome sequence is not read\n","\n","features_genes = features_SGD[features_SGD['feature'] == 'gene']\n","features_genes = features_genes.reset_index(drop=True)"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"jLGyWVkJhW3o"},"source":["# The features_genes file is used to build a dictionary of standard name -\n","# systematic name associations\n","\n","genes_orf = {}\n","\n","for row in range(features_genes.shape[0]):\n","\n","  features = features_genes.at[row, 'attribute'].split(';')\n","  feat_dict = {}\n","\n","  for feat in features:\n","    feat_split = feat.split('=')\n","\n","    feat_dict[feat_split[0]] = feat_split[1]\n","\n","  if 'gene' in feat_dict.keys(): # For cases when the two IDs are the same \n","                                 # (ex: mitochondrial genes)\n","\n","    genes_orf[feat_dict['gene']] = feat_dict['ID']\n","\n","  else:\n","\n","    genes_orf[feat_dict['ID']] = feat_dict['ID']"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"iCmJM_y4iJEY","executionInfo":{"status":"ok","timestamp":1635277147449,"user_tz":240,"elapsed":294,"user":{"displayName":"Simon Aubé","photoUrl":"https://lh3.googleusercontent.com/a/default-user=s64","userId":"10741788869764524869"}},"outputId":"a5622218-c466-4986-bd10-02cb912c86a0"},"source":["# This dictionary is used to add the systematic ID of genes to the rates df\n","rates_yeast.insert(0, 'ORF', 'problem')\n","missing = []\n","\n","for row in range(rates_yeast.shape[0]):\n","  gene = rates_yeast.at[row, 'Gene']\n","\n","  if gene in genes_orf.keys():\n","    rates_yeast.at[row, 'ORF'] = genes_orf[gene]\n","  \n","  else:\n","\n","    missing.append(gene)\n","\n","print(missing)"],"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["['AIM1', 'PHO88', 'DUR1,2', 'ARG5,6', 'FRA2', 'ADE5,7', 'AML1', 'YSP1', 'PET10', 'ENV10', 'MDN1', 'FLD1', 'OSW5', 'ICY2']\n"]}]},{"cell_type":"code","metadata":{"id":"0B5K0EYoiTqU"},"source":["# For the missing genes, corresponding to alternative aliases, the systematic \n","# IDs are added manually\n","\n","missing_ID = ['YAL046C', 'YBR106W', 'YBR208C', 'YER069W', 'YGL220W', 'YGL234W',\n","              'YGR001C', 'YHR155W', 'YKR046C', 'YLR065C', 'YLR106C', 'YLR404W',\n","              'YMR148W', 'YPL250C']\n","\n","count = 0\n","for gene in missing:\n","\n","  row = rates_yeast[rates_yeast['Gene'] == gene].index\n","\n","  rates_yeast.at[row, 'ORF'] = missing_ID[count]\n","\n","  count += 1"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"HL7zJb4BiwF7"},"source":["## 2) Addition of duplication-type annotations"]},{"cell_type":"markdown","metadata":{"id":"n4Hz9D7ki3eR"},"source":["Table S1 from Supplementary file 2 of Marchant et al. (2019) is imported. In this file, each yeast ORF (systematic ID) is associated to a duplication type annotation (S: singleton, SSD: small scale duplication,  SSD-successive: successive small-scale duplication, WGD: whole-genome duplication or SSD_WGD: paralogs duplicated by both mechanism of duplication SSD and WGD)."]},{"cell_type":"code","metadata":{"id":"LqwN8G2VjVeN"},"source":["dupli_type = pd.read_csv('/content/drive/My Drive/Redaction_SA/Figures_2021_06/Data_init/TableS1_Marchant_2019.csv', sep=';')"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"f09jlRF3lfso"},"source":["rates_yeast['Duplication'] = 'Unspecified' # For cases that are not in\n","                                           # the dupli_type dataframe\n","\n","for row in range(rates_yeast.shape[0]):\n","  orf = rates_yeast.at[row, 'ORF']\n","\n","  if orf in list(dupli_type['ORF']):\n","    dupli = dupli_type[dupli_type['ORF'] == orf]['Duplication'].values[0]\n","    rates_yeast.at[row, 'Duplication'] = dupli"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"g-ep7idVmbGC"},"source":["## 3) Recalculation of transcription and translation rates and rates_Hausser.csv export"]},{"cell_type":"markdown","source":["We recalculated all expression rates reported by Hausser et al. (2019) from the raw data deposited by Weinberg et al. This was done because we mistakenly thought that the calculation of translation rates by Hausser et al. was erroneous. "],"metadata":{"id":"Qtc3zEvzB6VE"}},{"cell_type":"markdown","source":["These recalculated rates are not used in our work, but some of the preliminary simulations were performed using both the original and recalculated (\"corrected\") sets of translations rates. In these, the choice of the dataset had no impact on the results."],"metadata":{"id":"I8rjttBvJgdz"}},{"cell_type":"markdown","metadata":{"id":"kkAvs6l0mp-K"},"source":["Transcription and translation rates, as well as protein abundance, are recalculated from the RPKMs reported by Weinberg et al. (2016), following the Methods of Hausser et al. (2019)."]},{"cell_type":"markdown","metadata":{"id":"gIz9PQpcrDNr"},"source":["Raw data are downloaded from GSE75897 (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE75897). The RNA-seq RPKM from the RiboZero experiment (GSE75897_RiboZero_RPKMs.txt.gz) and the ribosome profiling RPKMs (GSE75897_RPF_RPKMs.txt.gz) are obtained and imported:"]},{"cell_type":"code","metadata":{"id":"qIImz056nEYW"},"source":["mRNA_RPKM = pd.read_csv('/content/drive/My Drive/Redaction_SA/Figures_2021_06/Data_init/GSE75897_RiboZero_RPKMs.txt', sep='\\t', names=['ORF', 'mRNA RPKM'])\n","RP_RPKM = pd.read_csv('/content/drive/My Drive/Redaction_SA/Figures_2021_06/Data_init/GSE75897_RPF_RPKMs.txt', sep='\\t', names=['ORF', 'RP RPKM'])"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"xvyCQGOHsFHx"},"source":["# To obtain a file with only genes for which RPKMs have been reported in\n","# both experiments\n","\n","RPKM_all = pd.merge(mRNA_RPKM, RP_RPKM, how='outer')\n","\n","RPKM_both = RPKM_all.dropna().reset_index(drop=True)\n","\n","# RPKM of 0.0 are also dropped\n","RPKM_both = RPKM_both[(RPKM_both['mRNA RPKM'] != 0) & (RPKM_both['RP RPKM'] != 0)].reset_index(drop=True)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"IwI27q9Ksc1B"},"source":["There are 5068 genes for which both types of RPKMs are available, compared with 4400 in the dataset from Hausser et al. (2019). This seems consistent with the filtering step described in the Methods of this paper."]},{"cell_type":"code","metadata":{"id":"OUShMhMRib9i"},"source":["# Constants from Hausser et al. (2019)\n","alpha_m = 5.1\n","alpha_p = 1.34\n","n_mRNA = 60000\n","n_prot = 1.1e8\n","\n","# Transcription and translation rates are calculated (Hausser et al., 2019)\n","RPKM_both['m_est'] = np.log10(n_mRNA*(RPKM_both['mRNA RPKM']/np.sum(RPKM_both['mRNA RPKM'])))\n","RPKM_both['bm_calc'] = np.log10((10**RPKM_both['m_est']) * alpha_m)\n","\n","RPKM_both['bp_calc'] = np.log10(((n_prot * alpha_p)/(10**RPKM_both['m_est']))*((RPKM_both['RP RPKM'])/np.sum(RP_RPKM['RP RPKM'])))\n","\n","# As well as a new estimated protein abundance\n","RPKM_both['pEst_calc'] = np.log10(((10**RPKM_both['bm_calc']) * (10**RPKM_both['bp_calc'])) / (alpha_m * alpha_p))"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"tR9yJ8GsivHW"},"source":["# Merging with the yeastRates dataframe \n","rates_corrected = pd.merge(rates_yeast, RPKM_both[['ORF', 'm_est', 'bm_calc', 'bp_calc', 'pEst_calc']], how='left', on=['ORF'])"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"4T1TNA5BjePG"},"source":["This is the final rates_Hausser dataframe which was used to initialize the simulations. It is exported:"]},{"cell_type":"code","metadata":{"id":"zGXAbCDkjoCg"},"source":["rates_corrected.to_csv(path_or_buf='/content/drive/My Drive/Redaction_SA/Figures_2021_06/Data_sim_ready/rates_Hausser.csv', index=False, header=True)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"bLhBkngWksAw"},"source":["## 4) Reconstitution of paralog couples and calculation of relative divergence"]},{"cell_type":"markdown","metadata":{"id":"nq0yNQ8IpBeP"},"source":["The P1-P2 associations are taken from Table S9 of Supplementary File 2 from Marchant et al. (2019), which compiles both SSD- and WGD-derived paralog pairs (as well as other types of paralogs such as multi-SSD which are ignored in the current work). \n"]},{"cell_type":"code","metadata":{"id":"S0q5sXOCoRQS"},"source":["dupli_pairs = pd.read_csv('/content/drive/My Drive/Redaction_SA/Figures_2021_06/Data_init/TableS9_Marchant_2019.csv', sep=';')\n","dupli_pairs = dupli_pairs.dropna(subset=['Duplication', 'P1', 'P2']).reset_index(drop=True)"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"_hohIJCasWsN"},"source":["# A new dataframe is defined\n","rates_couples = pd.DataFrame(columns=['Duplication', 'ORF_P1', 'ORF_P2'])\n","\n","# After merging, the dataframe will have all those columns:\n","# Duplication: Duplication type from Marchant et al.\n","# ORF_P1: Systematic name paralog 1\n","# Gene_P1: Standard name paralog 1\n","# ORF_P2: Systematic name paralog 2\n","# Gene_P2: Standard name paralog 2\n","# Couple: P1-P1 (standard names)\n","# bm_P1: Log10 transcription rate of P1, in mRNAs/h (Hausser et al.)\n","# bm_P2: Log10 transcription rate of P2, in mRNAs/h (Hausser et al.)\n","# bp_P1: Log10 Translation rate of P1, in proteins per mRNA per hour (Hausser et al.)\n","# bp_calc_P1: Recalculated translation rate of P1\n","# bp_P2: Log10 Translation rate of P2, in proteins per mRNA per hour (Hausser et al.)\n","# bp_calc_P2: Recalculated translation rate of P2\n","# pEst_P1: Log10 Estimated protein abundance for P1 (proteins / cell)\n","# pEst_P2 : Log10 Estiimated protein abundance for P2 (proteins per cell)\n","# cv_P1: Coefficient of variation for P1 (Newman et al., 2006)\n","# cv_P2: Coefficient of variation for P2 (Newman et al., 2006)\n","\n","rates_couples['Duplication'] = dupli_pairs['Duplication']\n","rates_couples['ORF_P1'] = dupli_pairs['P1']\n","rates_couples['ORF_P2'] = dupli_pairs['P2']"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"aKQYl14-0Qte"},"source":["# New versions of the rates_corrected dataframe are prepared to be merged with\n","# the rates_couples df just created\n","\n","rates_only = rates_corrected[['ORF', 'Gene', 'bm', 'bp', 'cv', 'pEst',\n","                              'Duplication', 'bp_calc', 'pEst_calc']].copy()\n","\n","rates_P1 = rates_only.copy()\n","rates_P1 = rates_P1.rename(columns={'ORF': 'ORF_P1', 'Gene': 'Gene_P1', \n","                                    'bm': 'bm_P1', 'bp': 'bp_P1', 'cv': 'cv_P1',\n","                                    'pEst': 'pEst_P1', 'bp_calc': 'bp_calc_P1',\n","                                    'pEst_calc': 'pEst_calc_P1'})\n","\n","rates_P2 = rates_only.copy()\n","rates_P2 = rates_P2.rename(columns={'ORF': 'ORF_P2', 'Gene': 'Gene_P2', \n","                                    'bm': 'bm_P2', 'bp': 'bp_P2', 'cv': 'cv_P2',\n","                                    'pEst': 'pEst_P2', 'bp_calc': 'bp_calc_P2',\n","                                    'pEst_calc': 'pEst_calc_P2'})"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"KFWUpWLe1whp"},"source":["# The dataframe is then filled\n","rates_couples = pd.merge(rates_couples, rates_P1, how='left',\n","                         on=['Duplication', 'ORF_P1'])\n","\n","rates_couples = pd.merge(rates_couples, rates_P2, how='left',\n","                         on=['Duplication', 'ORF_P2'])"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"CQSSIYxWcaWN"},"source":["# Keeping only pairs where both paralogs were included in the dataset\n","rates_couples = rates_couples.dropna(subset=['bm_P1', 'bm_P2', 'bp_P1', 'bp_P2']).reset_index(drop=True)"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"nEgUx58-rsSO"},"source":["# Rearranging the columns in the desired order:\n","rates_couples = rates_couples[['Duplication', 'ORF_P1', 'Gene_P1', 'ORF_P2', \n","                               'Gene_P2', 'bm_P1', 'bm_P2', 'bp_P1', \n","                               'bp_calc_P1', 'bp_P2', 'bp_calc_P2', 'pEst_P1', \n","                               'pEst_calc_P1', 'pEst_P2', 'pEst_calc_P2', \n","                               'cv_P1', 'cv_P2']]"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"ywWhZ1-xvQWH"},"source":["# Definition of a function to compute log2 fold-changes of relative divergence\n","\n","def fold_change(prop_P1, prop_P2, data, name):\n","  \"\"\"Function to calculate log2 fold-change for a property between two\n","     duplicates. The two properties are provided as dataframe columns.\n","     The last argument is used to specify the name of the new column \n","     inserted in the data dataframe.\"\"\"\n","\n","  df = data.copy()\n","  df[name] = np.NaN\n","\n","  for row in range(df.shape[0]):\n","\n","    value_P1 = df.at[row, f'{prop_P1}']\n","    value_P2 = df.at[row, f'{prop_P2}']\n","\n","    if value_P1 >= value_P2:\n","\n","      df.at[row, name] = (10**value_P1)/(10**value_P2)\n","\n","    elif value_P1 < value_P2:\n","\n","      df.at[row, name] = (10**value_P2)/(10**value_P1)\n","\n","  df[name] = np.log2(df[name])\n","\n","  return df"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"LBNRt4DAv5co"},"source":["# Calculation of the log2 fold-changes of divergence\n","rates_logfold = fold_change('bm_P1', 'bm_P2', rates_couples, 'bm_fold_original')\n","rates_logfold = fold_change('bp_P1', 'bp_P2', rates_logfold, 'bp_fold_original')\n","rates_logfold = fold_change('pEst_P1', 'pEst_P2', rates_logfold, 'pEst_fold_original')\n","\n","rates_logfold = fold_change('bp_calc_P1', 'bp_calc_P2', rates_logfold, 'bp_fold_calc')\n","rates_logfold = fold_change('pEst_calc_P1', 'pEst_calc_P2', rates_logfold, 'pEst_fold_calc')"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"WwdOtbaJyYNQ"},"source":["# Calculation of normalized differences\n","rates_div_rel = rates_logfold.copy()\n","\n","rates_div_rel['bm_diff'] = np.abs(10**rates_div_rel['bm_P2'] - 10**rates_div_rel['bm_P1']) / (10**rates_div_rel['bm_P2'] + 10**rates_div_rel['bm_P1'])\n","rates_div_rel['bp_diff'] = np.abs(10**rates_div_rel['bp_P2'] - 10**rates_div_rel['bp_P1']) / (10**rates_div_rel['bp_P2'] + 10**rates_div_rel['bp_P1'])\n","rates_div_rel['pEst_diff'] = np.abs(10**rates_div_rel['pEst_P2'] - 10**rates_div_rel['pEst_P1']) / (10**rates_div_rel['pEst_P2'] + 10**rates_div_rel['pEst_P1'])\n","rates_div_rel['bp_calc_diff'] = np.abs(10**rates_div_rel['bp_calc_P2'] - 10**rates_div_rel['bp_calc_P1']) / (10**rates_div_rel['bp_calc_P2'] + 10**rates_div_rel['bp_calc_P1'])\n","rates_div_rel['pEst_calc_diff'] = np.abs(10**rates_div_rel['pEst_calc_P2'] - 10**rates_div_rel['pEst_calc_P1']) / (10**rates_div_rel['pEst_calc_P2'] + 10**rates_div_rel['pEst_calc_P1'])"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"HTMiEtAt0y40"},"source":["This is the final dataframe with log2 fold-changes and normalized differences that is used to generate figure 1 as well as to initialize the simulations (and compare the simulated and empirical data at the end). It is exported as csv:"]},{"cell_type":"code","metadata":{"id":"3YWb3Z7A1UxX"},"source":["rates_div_rel.to_csv(path_or_buf='/content/drive/My Drive/Redaction_SA/Figures_2021_06/Data_sim_ready/couples_divergence.csv', index=False, header=True)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"GopN80iYKqFA"},"source":["The empirical correlations of logfold divergence are calculated:"]},{"cell_type":"code","metadata":{"id":"sdYJ0LjHKxFL"},"source":["r_corr = stats.pearsonr(rates_div_rel['bm_fold_original'], rates_div_rel['bp_fold_original'])\n","r_calc = stats.pearsonr(rates_div_rel['bm_fold_original'], rates_div_rel['bp_fold_calc'])\n","\n","rho_corr = stats.spearmanr(rates_div_rel['bm_fold_original'], rates_div_rel['bp_fold_original'])\n","rho_calc = stats.spearmanr(rates_div_rel['bm_fold_original'], rates_div_rel['bp_fold_calc'])"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"oTr-z7FKLpXx","executionInfo":{"status":"ok","timestamp":1623450364881,"user_tz":240,"elapsed":2,"user":{"displayName":"Simon Aubé","photoUrl":"","userId":"10741788869764524869"}},"outputId":"5b127516-0ee3-4990-821a-b508e0b745c8"},"source":["r_corr"],"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["(0.26115963770858425, 8.38039843945512e-08)"]},"metadata":{"tags":[]},"execution_count":26}]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"mCZeAT_HLqdp","executionInfo":{"status":"ok","timestamp":1623450370055,"user_tz":240,"elapsed":2485,"user":{"displayName":"Simon Aubé","photoUrl":"","userId":"10741788869764524869"}},"outputId":"f8e45612-d2bf-48ca-af86-41616a7a215d"},"source":["r_calc"],"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["(0.28408216841747713, 4.9587573379420115e-09)"]},"metadata":{"tags":[]},"execution_count":27}]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"YKEA-tkzLs4B","executionInfo":{"status":"ok","timestamp":1623450379489,"user_tz":240,"elapsed":3,"user":{"displayName":"Simon Aubé","photoUrl":"","userId":"10741788869764524869"}},"outputId":"8cb013c5-dec5-4d8d-fbaa-a8404c4e8d65"},"source":["rho_corr"],"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["SpearmanrResult(correlation=0.2781015739828044, pvalue=1.0635357565012332e-08)"]},"metadata":{"tags":[]},"execution_count":28}]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"0NUALRzOLuJ6","executionInfo":{"status":"ok","timestamp":1623450389069,"user_tz":240,"elapsed":400,"user":{"displayName":"Simon Aubé","photoUrl":"","userId":"10741788869764524869"}},"outputId":"5657a202-8139-4cb1-8f6d-0b8b81195b4f"},"source":["rho_calc"],"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["SpearmanrResult(correlation=0.2916321433457395, pvalue=1.8437959466822292e-09)"]},"metadata":{"tags":[]},"execution_count":29}]},{"cell_type":"markdown","metadata":{"id":"rIXZmGvYM72Q"},"source":["Correlations with are also calculated from log2 fold-changes computed in randomized orientations (only for original data)"]},{"cell_type":"code","metadata":{"id":"VeJCMz6Fc67P"},"source":["rates_rand = rates_div_rel[['bm_P1', 'bm_P2', 'bp_P1', 'bp_P2']].copy()\n","rates_rand = 10**rates_rand\n","rng = np.random.default_rng()\n","corr_model = pd.DataFrame(columns=['pearson', 'r_pval', 'spearman', 'rho_pval'])\n","corr_rand = corr_model.copy()"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"JmsyId3JdV55"},"source":["for i in range(50):\n","  log_rand = rates_rand.copy()\n","  \n","  assos1 = rng.integers(0, 2, log_rand.shape[0])\n","  assos2 = np.abs(assos1 - 1)\n","  \n","  P1_rand = log_rand.copy()\n","  P2_rand = log_rand.copy()\n","  \n","  P1_rand['Logfold_Bm'] = P1_rand['bm_P1'] / P1_rand['bm_P2']\n","  P1_rand['Logfold_Bp'] = P1_rand['bp_P1'] / P1_rand['bp_P2']\n","  P2_rand['Logfold_Bm'] = P2_rand['bm_P2'] / P2_rand['bm_P1']\n","  P2_rand['Logfold_Bp'] = P2_rand['bp_P2'] / P2_rand['bp_P1']\n","  \n","  P1_rand = P1_rand * assos1[:, np.newaxis]\n","  P2_rand = P2_rand * assos2[:, np.newaxis]\n","\n","  log_rand = P1_rand.copy()\n","  \n","  log_rand = log_rand + P2_rand\n","  \n","  log_rand['Logfold_Bm'] = np.log2(log_rand['Logfold_Bm'])\n","  log_rand['Logfold_Bp'] = np.log2(log_rand['Logfold_Bp'])\n","\n","  corr_data = corr_model.copy()\n","  pearson = stats.pearsonr(log_rand['Logfold_Bm'], log_rand['Logfold_Bp'])\n","  spearman = stats.spearmanr(log_rand['Logfold_Bm'], log_rand['Logfold_Bp'])\n","\n","  corr_data.at[0, 'pearson'] = pearson[0]\n","  corr_data.at[0, 'r_pval'] = pearson[1]\n","  corr_data.at[0, 'spearman'] = spearman[0]\n","  corr_data.at[0, 'rho_pval'] = spearman[1]\n","\n","  corr_rand = pd.concat([corr_rand, corr_data])\n","\n","corr_rand = corr_rand.reset_index(drop=True)"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"tLK7dC5JlUJt","executionInfo":{"status":"ok","timestamp":1623451171760,"user_tz":240,"elapsed":407,"user":{"displayName":"Simon Aubé","photoUrl":"","userId":"10741788869764524869"}},"outputId":"5ccecf8d-236f-46a7-bf28-9fa6e58e8d54"},"source":["# Median pearson correlation coefficient:\n","corr_rand['pearson'].median()"],"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["0.43356900636988405"]},"metadata":{"tags":[]},"execution_count":33}]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"7B3ZlYYPluvz","executionInfo":{"status":"ok","timestamp":1623451174235,"user_tz":240,"elapsed":519,"user":{"displayName":"Simon Aubé","photoUrl":"","userId":"10741788869764524869"}},"outputId":"6026c097-5e8b-4239-d8d2-c7eeadf1db91"},"source":["# Median pearson correlation p-values:\n","corr_rand['r_pval'].median()"],"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["3.564185391413534e-20"]},"metadata":{"tags":[]},"execution_count":34}]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"0wiPi7oalasC","executionInfo":{"status":"ok","timestamp":1623451174643,"user_tz":240,"elapsed":413,"user":{"displayName":"Simon Aubé","photoUrl":"","userId":"10741788869764524869"}},"outputId":"6e8b4737-4d36-4365-97f9-ea9983919831"},"source":["# Median spearman correlation coefficient:\n","corr_rand['spearman'].median()"],"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["0.43367207697677435"]},"metadata":{"tags":[]},"execution_count":35}]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"aGG08Jwol6TM","executionInfo":{"status":"ok","timestamp":1623451174643,"user_tz":240,"elapsed":3,"user":{"displayName":"Simon Aubé","photoUrl":"","userId":"10741788869764524869"}},"outputId":"14748a56-1040-4f30-b288-e93a3cc49de0"},"source":["# Median spearman correlation p-values:\n","corr_rand['rho_pval'].median()"],"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["3.484423291682522e-20"]},"metadata":{"tags":[]},"execution_count":36}]},{"cell_type":"markdown","metadata":{"id":"WsXY5z3zPgPT"},"source":["Last thing: How often is the most expressed copy also the most transcribed / translated?"]},{"cell_type":"code","metadata":{"id":"c8q3cQovPrEr"},"source":["P1_most = np.where(rates_div_rel['pEst_P1'] > rates_div_rel['pEst_P2'], 1, 0)\n","P2_most = np.where(rates_div_rel['pEst_P2'] > rates_div_rel['pEst_P1'], 1, 0)\n","\n","bm_P1 = np.where(rates_div_rel['bm_P1'] >= rates_div_rel['bm_P2'], 1, 0) * P1_most\n","bm_P2 = np.where(rates_div_rel['bm_P2'] > rates_div_rel['bm_P1'], 1, 0) * P2_most\n","bm_freq = (np.sum(bm_P1) + np.sum(bm_P2)) / rates_div_rel.shape[0]\n","\n","bp_P1 = np.where(rates_div_rel['bp_P1'] >= rates_div_rel['bp_P2'], 1, 0) * P1_most\n","bp_P2 = np.where(rates_div_rel['bp_P2'] > rates_div_rel['bp_P1'], 1, 0) * P2_most\n","bp_freq = (np.sum(bp_P1) + np.sum(bp_P2)) / rates_div_rel.shape[0]"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"WmQCqZKMQo6W","executionInfo":{"status":"ok","timestamp":1623451720914,"user_tz":240,"elapsed":8,"user":{"displayName":"Simon Aubé","photoUrl":"","userId":"10741788869764524869"}},"outputId":"68e0541a-ab35-448a-b475-b7e46c2a5eef"},"source":["print(bm_freq, bp_freq)"],"execution_count":null,"outputs":[{"output_type":"stream","text":["0.863080684596577 0.7677261613691931\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":439},"id":"Z9NTfj84Pm4e","executionInfo":{"status":"ok","timestamp":1623451417224,"user_tz":240,"elapsed":334,"user":{"displayName":"Simon Aubé","photoUrl":"","userId":"10741788869764524869"}},"outputId":"071b189c-e9b6-4648-8a90-eed6dd43bd7f"},"source":["rates_div_rel"],"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>Duplication</th>\n","      <th>ORF_P1</th>\n","      <th>Gene_P1</th>\n","      <th>ORF_P2</th>\n","      <th>Gene_P2</th>\n","      <th>bm_P1</th>\n","      <th>bm_P2</th>\n","      <th>bp_P1</th>\n","      <th>bp_calc_P1</th>\n","      <th>bp_P2</th>\n","      <th>bp_calc_P2</th>\n","      <th>pEst_P1</th>\n","      <th>pEst_calc_P1</th>\n","      <th>pEst_P2</th>\n","      <th>pEst_calc_P2</th>\n","      <th>cv_P1</th>\n","      <th>cv_P2</th>\n","      <th>bm_fold_original</th>\n","      <th>bp_fold_original</th>\n","      <th>pEst_fold_original</th>\n","      <th>bp_fold_calc</th>\n","      <th>pEst_fold_calc</th>\n","      <th>bm_diff</th>\n","      <th>bp_diff</th>\n","      <th>pEst_diff</th>\n","      <th>bp_calc_diff</th>\n","      <th>pEst_calc_diff</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>SSD</td>\n","      <td>YAL036C</td>\n","      <td>RBG1</td>\n","      <td>YGR173W</td>\n","      <td>RBG2</td>\n","      <td>1.718675</td>\n","      <td>1.643553</td>\n","      <td>3.092909</td>\n","      <td>3.350873</td>\n","      <td>3.097578</td>\n","      <td>3.354366</td>\n","      <td>3.978390</td>\n","      <td>4.237767</td>\n","      <td>3.907936</td>\n","      <td>4.166139</td>\n","      <td>-0.852169</td>\n","      <td>-0.871762</td>\n","      <td>0.249550</td>\n","      <td>0.015510</td>\n","      <td>0.234040</td>\n","      <td>0.011605</td>\n","      <td>0.237945</td>\n","      <td>0.086272</td>\n","      <td>0.005375</td>\n","      <td>0.080935</td>\n","      <td>0.004022</td>\n","      <td>0.082279</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>SSD</td>\n","      <td>YAL039C</td>\n","      <td>CYC3</td>\n","      <td>YKL087C</td>\n","      <td>CYT2</td>\n","      <td>0.891671</td>\n","      <td>1.082354</td>\n","      <td>3.449659</td>\n","      <td>3.570784</td>\n","      <td>3.456405</td>\n","      <td>3.498349</td>\n","      <td>3.508135</td>\n","      <td>3.630675</td>\n","      <td>3.705564</td>\n","      <td>3.748922</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>0.633434</td>\n","      <td>0.022409</td>\n","      <td>0.655843</td>\n","      <td>0.240625</td>\n","      <td>0.392809</td>\n","      <td>0.216071</td>\n","      <td>0.007766</td>\n","      <td>0.223463</td>\n","      <td>0.083201</td>\n","      <td>0.135302</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>SSD</td>\n","      <td>YAL042W</td>\n","      <td>ERV46</td>\n","      <td>YML067C</td>\n","      <td>ERV41</td>\n","      <td>1.970836</td>\n","      <td>1.617977</td>\n","      <td>2.782261</td>\n","      <td>3.091116</td>\n","      <td>2.991756</td>\n","      <td>3.229293</td>\n","      <td>3.919902</td>\n","      <td>4.230172</td>\n","      <td>3.776538</td>\n","      <td>4.015489</td>\n","      <td>NaN</td>\n","      <td>-0.666150</td>\n","      <td>1.172172</td>\n","      <td>0.695928</td>\n","      <td>0.476245</td>\n","      <td>0.459012</td>\n","      <td>0.713160</td>\n","      <td>0.385279</td>\n","      <td>0.236620</td>\n","      <td>0.163571</td>\n","      <td>0.157753</td>\n","      <td>0.242249</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>SSD</td>\n","      <td>YAL060W</td>\n","      <td>BDH1</td>\n","      <td>YAL061W</td>\n","      <td>BDH2</td>\n","      <td>1.823724</td>\n","      <td>0.887096</td>\n","      <td>2.972863</td>\n","      <td>3.245824</td>\n","      <td>3.073859</td>\n","      <td>3.384797</td>\n","      <td>3.963392</td>\n","      <td>4.237767</td>\n","      <td>3.127760</td>\n","      <td>3.440113</td>\n","      <td>-0.580871</td>\n","      <td>NaN</td>\n","      <td>3.111409</td>\n","      <td>0.335500</td>\n","      <td>2.775910</td>\n","      <td>0.461658</td>\n","      <td>2.649751</td>\n","      <td>0.792580</td>\n","      <td>0.115754</td>\n","      <td>0.745193</td>\n","      <td>0.158647</td>\n","      <td>0.725108</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>SSD</td>\n","      <td>YBL002W</td>\n","      <td>HTB2</td>\n","      <td>YDR224C</td>\n","      <td>HTB1</td>\n","      <td>2.552705</td>\n","      <td>2.507437</td>\n","      <td>3.724892</td>\n","      <td>3.535228</td>\n","      <td>3.833681</td>\n","      <td>3.644017</td>\n","      <td>5.444402</td>\n","      <td>5.256152</td>\n","      <td>5.507923</td>\n","      <td>5.319673</td>\n","      <td>-0.951170</td>\n","      <td>NaN</td>\n","      <td>0.150378</td>\n","      <td>0.361389</td>\n","      <td>0.211011</td>\n","      <td>0.361389</td>\n","      <td>0.211011</td>\n","      <td>0.052070</td>\n","      <td>0.124597</td>\n","      <td>0.073001</td>\n","      <td>0.124597</td>\n","      <td>0.073001</td>\n","    </tr>\n","    <tr>\n","      <th>...</th>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","      <td>...</td>\n","    </tr>\n","    <tr>\n","      <th>404</th>\n","      <td>WGD</td>\n","      <td>YIL135C</td>\n","      <td>VHS2</td>\n","      <td>YNL074C</td>\n","      <td>MLF3</td>\n","      <td>1.082863</td>\n","      <td>1.264818</td>\n","      <td>2.931731</td>\n","      <td>3.261975</td>\n","      <td>2.957714</td>\n","      <td>3.303574</td>\n","      <td>3.181400</td>\n","      <td>3.513058</td>\n","      <td>3.389337</td>\n","      <td>3.736611</td>\n","      <td>-0.666351</td>\n","      <td>NaN</td>\n","      <td>0.604439</td>\n","      <td>0.086313</td>\n","      <td>0.690751</td>\n","      <td>0.138190</td>\n","      <td>0.742629</td>\n","      <td>0.206471</td>\n","      <td>0.029905</td>\n","      <td>0.234925</td>\n","      <td>0.047857</td>\n","      <td>0.251839</td>\n","    </tr>\n","    <tr>\n","      <th>405</th>\n","      <td>WGD</td>\n","      <td>YMR177W</td>\n","      <td>MMT1</td>\n","      <td>YPL224C</td>\n","      <td>MMT2</td>\n","      <td>0.959662</td>\n","      <td>1.090190</td>\n","      <td>2.711851</td>\n","      <td>3.110034</td>\n","      <td>2.656638</td>\n","      <td>3.032142</td>\n","      <td>2.838319</td>\n","      <td>3.237916</td>\n","      <td>2.913634</td>\n","      <td>3.290551</td>\n","      <td>NaN</td>\n","      <td>NaN</td>\n","      <td>0.433605</td>\n","      <td>0.183414</td>\n","      <td>0.250190</td>\n","      <td>0.258753</td>\n","      <td>0.174852</td>\n","      <td>0.149155</td>\n","      <td>0.063481</td>\n","      <td>0.086493</td>\n","      <td>0.089437</td>\n","      <td>0.060525</td>\n","    </tr>\n","    <tr>\n","      <th>406</th>\n","      <td>WGD</td>\n","      <td>YOR237W</td>\n","      <td>HES1</td>\n","      <td>YPL145C</td>\n","      <td>KES1</td>\n","      <td>0.238656</td>\n","      <td>1.932001</td>\n","      <td>2.694922</td>\n","      <td>3.023173</td>\n","      <td>2.939815</td>\n","      <td>3.268066</td>\n","      <td>2.100384</td>\n","      <td>2.430049</td>\n","      <td>4.038621</td>\n","      <td>4.368286</td>\n","      <td>NaN</td>\n","      <td>-0.920276</td>\n","      <td>5.625170</td>\n","      <td>0.813515</td>\n","      <td>6.438685</td>\n","      <td>0.813515</td>\n","      <td>6.438685</td>\n","      <td>0.960283</td>\n","      <td>0.274702</td>\n","      <td>0.977206</td>\n","      <td>0.274702</td>\n","      <td>0.977206</td>\n","    </tr>\n","    <tr>\n","      <th>407</th>\n","      <td>WGD</td>\n","      <td>YGR180C</td>\n","      <td>RNR4</td>\n","      <td>YJL026W</td>\n","      <td>RNR2</td>\n","      <td>2.387377</td>\n","      <td>2.181690</td>\n","      <td>3.213313</td>\n","      <td>3.442151</td>\n","      <td>3.102431</td>\n","      <td>3.394253</td>\n","      <td>4.767495</td>\n","      <td>4.997747</td>\n","      <td>4.450927</td>\n","      <td>4.744163</td>\n","      <td>-0.908685</td>\n","      <td>-0.942334</td>\n","      <td>0.683274</td>\n","      <td>0.368342</td>\n","      <td>1.051616</td>\n","      <td>0.159114</td>\n","      <td>0.842388</td>\n","      <td>0.232476</td>\n","      <td>0.126969</td>\n","      <td>0.349139</td>\n","      <td>0.055089</td>\n","      <td>0.283928</td>\n","    </tr>\n","    <tr>\n","      <th>408</th>\n","      <td>WGD</td>\n","      <td>YFL004W</td>\n","      <td>VTC2</td>\n","      <td>YPL019C</td>\n","      <td>VTC3</td>\n","      <td>2.007437</td>\n","      <td>2.322162</td>\n","      <td>2.530392</td>\n","      <td>3.138708</td>\n","      <td>2.522320</td>\n","      <td>3.134289</td>\n","      <td>3.704634</td>\n","      <td>4.314365</td>\n","      <td>4.011288</td>\n","      <td>4.624670</td>\n","      <td>-0.694541</td>\n","      <td>NaN</td>\n","      <td>1.045494</td>\n","      <td>0.026813</td>\n","      <td>1.018680</td>\n","      <td>0.014683</td>\n","      <td>1.030811</td>\n","      <td>0.347274</td>\n","      <td>0.009293</td>\n","      <td>0.339076</td>\n","      <td>0.005089</td>\n","      <td>0.342791</td>\n","    </tr>\n","  </tbody>\n","</table>\n","<p>409 rows × 27 columns</p>\n","</div>"],"text/plain":["    Duplication   ORF_P1 Gene_P1  ... pEst_diff bp_calc_diff  pEst_calc_diff\n","0           SSD  YAL036C    RBG1  ...  0.080935     0.004022        0.082279\n","1           SSD  YAL039C    CYC3  ...  0.223463     0.083201        0.135302\n","2           SSD  YAL042W   ERV46  ...  0.163571     0.157753        0.242249\n","3           SSD  YAL060W    BDH1  ...  0.745193     0.158647        0.725108\n","4           SSD  YBL002W    HTB2  ...  0.073001     0.124597        0.073001\n","..          ...      ...     ...  ...       ...          ...             ...\n","404         WGD  YIL135C    VHS2  ...  0.234925     0.047857        0.251839\n","405         WGD  YMR177W    MMT1  ...  0.086493     0.089437        0.060525\n","406         WGD  YOR237W    HES1  ...  0.977206     0.274702        0.977206\n","407         WGD  YGR180C    RNR4  ...  0.349139     0.055089        0.283928\n","408         WGD  YFL004W    VTC2  ...  0.339076     0.005089        0.342791\n","\n","[409 rows x 27 columns]"]},"metadata":{"tags":[]},"execution_count":38}]}]}